---
title: "EWAS data preparation"
author: "Yunfeng Liu"
date: "5/2/2022"
output: html_document

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```{r}
#set library path.
.libPaths("~/researchdrive/yliu/BIOS_all/Functional interpretation of age-related X methylation/Rlibs")
```


```{r}
#Load all necessary libraries.
library(BBMRIomics)
library(tidyverse)
library(DNAmArray)
library(minfi)
library(irlba)
library(ggfortify)
library(data.table)

#View available datasets.
data(package="BBMRIomics")

#Load DNA-methylation data (large, takes about 30 minutes).
bbmri.data(methData_Mvalues_BIOS_Freeze2_unrelated)
dim(mvalues)   # 481388   4386

# Load maskProbes data
maskProbes <- as.data.frame(fread("./Input/Input_01_HM450.hg19.manifest.tsv"))
```


```{r}
#Keep sex chromosome for the DNA-methylation data.
mvalues <- keepSeqlevels(mvalues, c("chrX", "chrY"), pruning.mode = "coarse")
dim(mvalues)  # 11177 (11130X+47Y)

## remove unreliable probes 
## hg19 genome:60466 probes on whole genome are recommended to remove based on zhou et.al

maskProbes <- filter(maskProbes,MASK_general=="TRUE")
mvalues <- mvalues[!(rownames(mvalues) %in% maskProbes$probeID),] 
dim(mvalues) # 9783   4386
```


```{r}
# Remove some samples (N=263) who their age is missing
table(is.na(mvalues$sampling_age))  # 263
col.mvalues<-as.data.frame(colData(mvalues))
col.mvalues<-col.mvalues[!is.na(col.mvalues$sampling_age),]
mvalues<-mvalues[,colnames(mvalues) %in% rownames(col.mvalues)]
dim(mvalues)  #9783 4123

#Remove problem samples.
#There are 1 covariate which contain certain levels causing singularities if they are included into the model:
#sample_plate: levels "OV0192DNA001" and "OV0192DNA002".
idx <- which(mvalues$sample_plate == "OV0192DNA001" | mvalues$sample_plate == "OV0192DNA002")

#Now remove them.
mvalues <- mvalues[,-idx]
dim(mvalues)  # 9783 4044

# Sex mislabeled issues (Unfortunately,some samples are mixed-up in sex,we should exclude them from this analysis)
# Because we don't know if their age are still correct
# Load beta values
bbmri.data(methData_Betas_BIOS_Freeze2_unrelated)
#Keep sex chromosome for the DNA-methylation data.
betas_chrX <- keepSeqlevels(betas, c("chrX"), pruning.mode = "coarse")

## remove unreliable probes
betas_chrX <- betas_chrX[!(rownames(betas_chrX) %in% maskProbes$probeID),] 

# Select same samples in betas.
betas_chrX<-betas_chrX[,colnames(mvalues)]
dim(betas_chrX) # 9777 4044

# Check if predicted and real sexes are identical
predictedSex <- as.data.frame(getSex.DNAmArray(assay(betas)))
colnames(predictedSex)<-"predictedSex"
predictedSex$predictedSex[grepl("Male",predictedSex$predictedSex)]<-"male"
predictedSex$predictedSex[grepl("Female",predictedSex$predictedSex)]<-"female"
mvalues$predictedSex<-predictedSex$predictedSex
mvalues$correctSex<-ifelse(colData(mvalues)$predictedSex==colData(mvalues)$sex,TRUE,FALSE)
table(colData(mvalues)$correctSex) # FALSE:11  TRUE:4032 NA:2(sex is missing)

col.mvalues<-filter(col.mvalues,correctSex==TRUE)
mvalues<-mvalues[,rownames(col.mvalues)]
betas_chrX<-betas_chrX[,rownames(col.mvalues)]
dim(mvalues) # 9783(X+Y) 4032
dim(betas_chrX) # 9777 4032

### One female sample should be removed based on PCA.
pc_betas_chrX<- prcomp_irlba(t(assay(betas_chrX)))
summary(pc_betas_chrX)
tiff("PCA plot for sex.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
autoplot(pc_betas_chrX, data=as.data.frame(colData(betas_chrX)), colour="sex",main="Principal components plot")
dev.off() 

# Find this "female" samples
# Spilited samples into men and women separately
col.mvalues_females<-filter(col.mvalues,sex=="female")
col.mvalues_males<-filter(col.mvalues,sex=="male")
betas_Xfemales<-betas_chrX[,rownames(col.mvalues_females)]
betas_Xmales<-betas_chrX[,rownames(col.mvalues_males)] 
dim(betas_Xfemales)    #  9777 2344
dim(betas_Xmales)      #  9777 1688

# Calculate the mean methylation of each female samples. 
# we found "BIOS691E2BA5" sample have lowest DNAm level,remove it.
betas_Xfemales_mean<-as.data.frame(rowMeans(t(assay(betas_Xfemales)),na.rm = TRUE))
colnames(betas_Xfemales_mean)<-"Mean methylation"
betas_Xfemales<-betas_Xfemales[,!(colnames(betas_Xfemales)%in%"BIOS691E2BA5")]
dim(betas_Xfemales)    #  9777 2343

### Density plot of female X-methylation and male X-methylation.
tiff("Females X methylation.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
densityPlot(assay(betas_Xfemales),main="Females X methylation", xlab="Beta values",xlim=c(0,1), cex.main=1.5, cex.lab=1.3, cex.axis=1.25)
dev.off()

tiff("Males X methylation.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
densityPlot(assay(betas_Xmales),main="Males X methylation", xlab="Beta values",xlim=c(0,1), cex.main=1.5, cex.lab=1.3, cex.axis=1.25)
dev.off()

# Similarly,remove this sample from betas and mvalues object.
mvalues<-mvalues[,!(colnames(mvalues)%in%"BIOS691E2BA5")]
betas_chrX<-betas_chrX[,!(colnames(betas_chrX)%in%"BIOS691E2BA5")]
col.mvalues<-col.mvalues[!(rownames(col.mvalues)%in%"BIOS691E2BA5"),]
dim(mvalues) # 9783 4031
dim(col.mvalues) # 4031 84
dim(betas_chrX) # 9777 4031
dim(betas_Xfemales)    #  9777 2343
dim(betas_Xmales)      #  9777 1688

#Save betas,betas_Xfemales and betas_Xfemales object.
save(betas_chrX,betas_Xfemales,betas_Xmales,file = "./Output/Output_01/Output_01_chrX_betas.RData")

# Calculate the mean methylation of X CpGs on samples
dat_mean_Xfemales<-as.data.frame(rowMeans(assay(betas_Xfemales),na.rm = TRUE))
colnames(dat_mean_Xfemales)<-"females"
dat_mean_Xmales<-as.data.frame(rowMeans(assay(betas_Xmales),na.rm = TRUE))
colnames(dat_mean_Xmales)<-"males"
dat_mean_X<-cbind(dat_mean_Xfemales,dat_mean_Xmales)
save(dat_mean_Xfemales,dat_mean_Xmales,dat_mean_X,file="dat_mean_X.RData")
```


```{r}
# Predicted Cell counts for all BIOS data by minfi
## Creat a object called RGset_female
rownames(samplesheet)<-make.names(samplesheet$uuid, unique=TRUE)
id<-intersect(colnames(mvalues),rownames(samplesheet)) # 4031
samplesheet<-samplesheet[id,]
mvalues<-mvalues[,id]

## Spilit women and men separately
mvalues_females<-mvalues[,na.omit(match(rownames(col.mvalues_females),colnames(mvalues)))] 
dim(mvalues_females)    # 9783 2343

mvalues_males<-mvalues[,na.omit(match(rownames(col.mvalues_males),colnames(mvalues)))]     
dim(mvalues_males)      # 9783 1688
samplesheet_female<-filter(samplesheet,sex=="female")  # 2343
samplesheet_male<-filter(samplesheet,sex=="male")      # 1688

## loading IDAT files
# women
library(BiocParallel)
register(MulticoreParam(10, log=TRUE))
RGset_female <- read.metharray.exp(targets = samplesheet_female)

Cellcounts_female<-estimateCellCounts(RGset_female,referencePlatform = "IlluminaHumanMethylation450k")
rownames(Cellcounts_female)<-make.names(rownames(samplesheet_female), unique=TRUE)
Cellcounts_female<-as.data.frame(Cellcounts_female)
save(Cellcounts_female,file = "./Output_01_Cellcounts_female.RData")
Cellcounts_female<-Cellcounts_female[colnames(mvalues_females),]

#Add additional variables to mvalues_females.
mvalues_females$CD8T_predicted <- Cellcounts_female$CD8T
mvalues_females$CD4T_predicted <- Cellcounts_female$CD4T
mvalues_females$NK_predicted <- Cellcounts_female$NK
mvalues_females$Bcell_predicted <- Cellcounts_female$Bcell
mvalues_females$Mono_predicted <- Cellcounts_female$Mono
mvalues_females$Gran_predicted <- Cellcounts_female$Gran

# Men
RGset_male <- read.metharray.exp(targets = samplesheet_male)
Cellcounts_male<-estimateCellCounts(RGset_male,referencePlatform = "IlluminaHumanMethylation450k") 
rownames(Cellcounts_male)<-make.names(rownames(samplesheet_male), unique=TRUE) 
Cellcounts_male<-as.data.frame(Cellcounts_male) 
save(Cellcounts_male,file = "./Output_01_Cellcounts_male.RData")  
Cellcounts_male<-Cellcounts_male[colnames(mvalues_males),]

#Add additional variables to mvalues_males.
mvalues_males$CD8T_predicted <- Cellcounts_male$CD8T
mvalues_males$CD4T_predicted <- Cellcounts_male$CD4T
mvalues_males$NK_predicted <- Cellcounts_male$NK
mvalues_males$Bcell_predicted <- Cellcounts_male$Bcell
mvalues_males$Mono_predicted <- Cellcounts_male$Mono
mvalues_males$Gran_predicted <- Cellcounts_male$Gran
```


```{r}
## keep X probes in females  
mvalues_Xfemales <- mvalues_females[seqnames(mvalues_females) %in% "chrX",]
dim(mvalues_Xfemales) #  9777  2343

## keep X probes in males  
mvalues_Xmales <- mvalues_males[seqnames(mvalues_males) %in% "chrX",]
dim(mvalues_Xmales) #  9777  1688

## keep Y probes in males  
mvalues_Ymales <- mvalues_males[seqnames(mvalues_males) %in% "chrY",]
dim(mvalues_Ymales) #  6  1688

#Save the files.
save(mvalues,mvalues_males,mvalues_Xfemales,mvalues_Xmales,mvalues_Ymales,file = "./Output/Output_01/Output_01_chrX_chrY.RData")
```

