---
title: "aVMCs validation"
author: "Yunfeng Liu"
date: "5/24/2022"
output: html_document
---

Here,we would like to replicate our aDMCs analysis on X chromosome using three external datasets.

1.Blood DNAm data generated by Hannum et al.: "Genome-wide Methylation Profiles Reveal Quantitative Views of Human Aging Rates.Mol Cell."
GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40279

2.Blood DNAm data generated by Johansson et al.: "Continuous Aging of the Human DNA Methylome Throughout the Human Lifespan.PLoS One."
GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87571

3.Monocytes DNAm data generated by Liu et al.: "Age-related variations in the methylome associated with gene expression in human monocytes and T cells. Nat Commun."
GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE56046

```{r}
#set library path.
.libPaths("~/researchdrive/yliu/BIOS_all/Functional interpretation of age-related X methylation/Rlibs")
```

```{r}
# Load data
load(file = "Output_05_DGLM_Hannum_Blood_Xfemales_converge.RData")
load(file = "Output_05_DGLM_Hannum_Blood_Xmales_converge.RData")
load(file = "Output_05_DGLM_Johansson_Blood_Xfemales_converge.RData")
load(file = "Output_05_DGLM_Johansson_Blood_Xmales_converge.RData")
load(file = "Output_05_DGLM_Liu_Mono_Xfemales_converge.RData")
load(file = "Output_05_DGLM_Liu_Mono_Xmales_converge.RData")
load(file = "Output_02_aDMCs_effectsize_BIOS.RData")
```

```{r}
#### aVMCs validation on Hannum blood datasets
## Correct bias and inflation for dispersion model.
# Xfemales
set.seed(1)
bc_DGLM_Hannum_Blood_Xfemales<- bacon(DGLM_Hannum_Blood_Xfemales_converge[,7])

# Calculate bias and inflation
bias(bc_DGLM_Hannum_Blood_Xfemales)         # 0.12  
inflation(bc_DGLM_Hannum_Blood_Xfemales)    # 0.9

# Save histogram plot of test statistics 
tiff("Output_06_Bacon_DGLM_Hannum_Blood_Xfemales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Hannum_Blood_Xfemales,xlab="t-statistics",main="aVMCs in Hannum Blood females",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Hannum_Blood_Xfemales<-tstat(bc_DGLM_Hannum_Blood_Xfemales)
pvals_DGLM_Hannum_Blood_Xfemales <-pval(bc_DGLM_Hannum_Blood_Xfemales)
table(pvals_DGLM_Hannum_Blood_Xfemales<0.05)   # 5598  
rownames(pvals_DGLM_Hannum_Blood_Xfemales)<-make.names(rownames(DGLM_Hannum_Blood_Xfemales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Hannum_Blood_Xfemales_es<-bacon(NULL,DGLM_Hannum_Blood_Xfemales_converge[,5],DGLM_Hannum_Blood_Xfemales_converge[,6])
es_DGLM_Hannum_Blood_Xfemales<-es(bc_DGLM_Hannum_Blood_Xfemales_es)

# Summary Bacon statistics
dat_DGLM_Hannum_Blood_Xfemales_bacon_dispersion<-as.data.frame(cbind(es_DGLM_Hannum_Blood_Xfemales,tstats_DGLM_Hannum_Blood_Xfemales,pvals_DGLM_Hannum_Blood_Xfemales))
colnames(dat_DGLM_Hannum_Blood_Xfemales_bacon_dispersion)<-c("bacon_effectsize_dispersion","bacon_tstatistics_dispersion","bacon_pval_dispersion")
rownames(dat_DGLM_Hannum_Blood_Xfemales_bacon_dispersion)<-make.names(rownames(DGLM_Hannum_Blood_Xfemales_converge),unique = TRUE)

# check females effecsize of aVMCs between BIOS and Hannum 
dat_Hannum_Blood_Xfemales_aVMCs<-dat_DGLM_Hannum_Blood_Xfemales_bacon_dispersion[rownames(df_aVMCs_effectsize_BIOS),]
df_Hannum_validation_females<-cbind(dat_Hannum_Blood_Xfemales_aVMCs[,1],df_aVMCs_effectsize_BIOS[,-2])
colnames(df_Hannum_validation_females)<-c("Hannum","BIOS","symbol")

tiff("Output_06_Hannum_Validation of female effect size.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Hannum_validation_females) +
    aes(x = BIOS,y = Hannum,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Hannum Blood")
dev.off()

# Xmales
set.seed(1)
bc_DGLM_Hannum_Blood_Xmales<- bacon(DGLM_Hannum_Blood_Xmales_converge[,7])

# Calculate bias and inflation
bias(bc_DGLM_Hannum_Blood_Xmales)         # 0.22  
inflation(bc_DGLM_Hannum_Blood_Xmales)    # 0.94

# Save histogram plot of test statistics 
tiff("Output_06_Bacon_DGLM_Hannum_Blood_Xmales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Hannum_Blood_Xmales,xlab="t-statistics",main="aVMCs in Hannum Blood males",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Hannum_Blood_Xmales<-tstat(bc_DGLM_Hannum_Blood_Xmales)
pvals_DGLM_Hannum_Blood_Xmales <-pval(bc_DGLM_Hannum_Blood_Xmales)
table(pvals_DGLM_Hannum_Blood_Xmales<0.05)   # 1333  
rownames(pvals_DGLM_Hannum_Blood_Xmales)<-make.names(rownames(DGLM_Hannum_Blood_Xmales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Hannum_Blood_Xmales_es<-bacon(NULL,DGLM_Hannum_Blood_Xmales_converge[,5],DGLM_Hannum_Blood_Xmales_converge[,6])
es_DGLM_Hannum_Blood_Xmales<-es(bc_DGLM_Hannum_Blood_Xmales_es)

# Summary Bacon statistics
dat_DGLM_Hannum_Blood_Xmales_bacon_dispersion<-as.data.frame(cbind(es_DGLM_Hannum_Blood_Xmales,tstats_DGLM_Hannum_Blood_Xmales,pvals_DGLM_Hannum_Blood_Xmales))
colnames(dat_DGLM_Hannum_Blood_Xmales_bacon_dispersion)<-c("bacon_effectsize_dispersion","bacon_tstatistics_dispersion","bacon_pval_dispersion")
rownames(dat_DGLM_Hannum_Blood_Xmales_bacon_dispersion)<-make.names(rownames(DGLM_Hannum_Blood_Xmales_converge),unique = TRUE)

# check males effecsize of aVMCs between BIOS and Hannum 
dat_Hannum_Blood_Xmales_aVMCs<-dat_DGLM_Hannum_Blood_Xmales_bacon_dispersion[rownames(df_aVMCs_effectsize_BIOS),]
df_Hannum_validation_males<-cbind(dat_Hannum_Blood_Xmales_aVMCs[,1],df_aVMCs_effectsize_BIOS[,-1])
colnames(df_Hannum_validation_males)<-c("Hannum","BIOS","symbol")

tiff("Output_06_Hannum_Validation of male effect size.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Hannum_validation_males) +
    aes(x = BIOS,y = Hannum,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Hannum Blood")
dev.off()
```

```{r}
# save the validation data of effect size 
save(df_Hannum_validation_females,df_Hannum_validation_males,file="Output_06_Hannum_validation.RData")
```

```{r}
# Extract 631 female p-values and 36 male p-values.
dat_Hannum_Blood_Xfemales_aVMCs$symbol<-df_aVMCs_effectsize_BIOS$symbol
df_aVMCs_females_Hannum_Blood<-filter(dat_Hannum_Blood_Xfemales_aVMCs,symbol=="female-specific")
pvals_aVMCs_Xfemales_Hannum_Blood<-as.data.frame(pvals_DGLM_Hannum_Blood_Xfemales[rownames(df_aVMCs_females_Hannum_Blood),])
colnames(pvals_aVMCs_Xfemales_Hannum_Blood)<-"pval_females"

dat_Hannum_Blood_Xmales_aVMCs$symbol<-df_aVMCs_effectsize_BIOS$symbol
df_aVMCs_males_Hannum_Blood<-filter(dat_Hannum_Blood_Xmales_aVMCs,symbol=="male-specific")
pvals_aVMCs_Xmales_Hannum_Blood<-as.data.frame(pvals_DGLM_Hannum_Blood_Xmales[rownames(df_aVMCs_males_Hannum_Blood),])
colnames(pvals_aVMCs_Xmales_Hannum_Blood)<-"pval_males"

# Perform FDR correction for 667 pvals(631+36)
padjs_aVMCs_Hannum_Blood<-as.data.frame(p.adjust(c(pvals_aVMCs_Xfemales_Hannum_Blood$pval_females,pvals_aVMCs_Xmales_Hannum_Blood$pval_males),method="BH"))
colnames(padjs_aVMCs_Hannum_Blood)<-"bacon_p.adj_dispersion"

# check females (how many of 631) and males (how many of 36) separately how many were FDR significant 
# 574 female aVMCs are still FDR significant in Hannum blood data.
padjs_aVMCs_females_Hannum_Blood<-as.data.frame(padjs_aVMCs_Hannum_Blood[1:631,])
colnames(padjs_aVMCs_females_Hannum_Blood)<-"bacon_p.adj_dispersion"
df_aVMCs_females_Hannum_Blood$bacon_p.adj_dispersion<-padjs_aVMCs_females_Hannum_Blood$bacon_p.adj_dispersion

df_aVMCs_females_Hannum_Blood<-filter(df_aVMCs_females_Hannum_Blood,bacon_p.adj_dispersion<0.05)
df_aVMCs_females_effectsize_up_Hannum<-filter(df_aVMCs_females_Hannum_Blood,bacon_effectsize_dispersion>0) # 574
df_aVMCs_females_effectsize_down_Hannum<-filter(df_aVMCs_females_Hannum_Blood,bacon_effectsize_dispersion<0) # 0

# 35 male aVMCs are still FDR significant in Hannum blood data.
padjs_aVMCs_males_Hannum_Blood<-as.data.frame(padjs_aVMCs_Hannum_Blood[632:667,])
colnames(padjs_aVMCs_males_Hannum_Blood)<-"bacon_p.adj_dispersion"
df_aVMCs_males_Hannum_Blood$bacon_p.adj_dispersion<-padjs_aVMCs_males_Hannum_Blood$bacon_p.adj_dispersion

df_aVMCs_males_Hannum_Blood<-filter(df_aVMCs_males_Hannum_Blood,bacon_p.adj_dispersion<0.05)
df_aVMCs_males_effectsize_up_Hannum<-filter(df_aVMCs_males_Hannum_Blood,bacon_effectsize_dispersion>0) # 34
df_aVMCs_males_effectsize_down_Hannum<-filter(df_aVMCs_males_Hannum_Blood,bacon_effectsize_dispersion<0) # 1

save(df_aVMCs_females_effectsize_up_Hannum,df_aVMCs_females_Hannum_Blood,df_aVMCs_males_effectsize_up_Hannum,df_aVMCs_males_effectsize_down_Hannum,df_aVMCs_males_Hannum_Blood,file = "Output_06_Hannum_Catogue of aVMCs.RData")
```

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```{r}
#### aVMCs validation on Johansson blood datasets
## Correct bias and inflation for dispersion model.
# Xfemales
set.seed(1)
bc_DGLM_Johansson_Blood_Xfemales<- bacon(DGLM_Johansson_Blood_Xfemales_converge[,7])

# Calculate bias and inflation
bias(bc_DGLM_Johansson_Blood_Xfemales)         # 0.14  
inflation(bc_DGLM_Johansson_Blood_Xfemales)    # 0.74

# Save histogram plot of test statistics 
tiff("Output_06_Bacon_DGLM_Johansson_Blood_Xfemales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Johansson_Blood_Xfemales,xlab="t-statistics",main="aVMCs in Johansson Blood females",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Johansson_Blood_Xfemales<-tstat(bc_DGLM_Johansson_Blood_Xfemales)
pvals_DGLM_Johansson_Blood_Xfemales <-pval(bc_DGLM_Johansson_Blood_Xfemales)
table(pvals_DGLM_Johansson_Blood_Xfemales<0.05)   # 6179  
rownames(pvals_DGLM_Johansson_Blood_Xfemales)<-make.names(rownames(DGLM_Johansson_Blood_Xfemales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Johansson_Blood_Xfemales_es<-bacon(NULL,DGLM_Johansson_Blood_Xfemales_converge[,5],DGLM_Johansson_Blood_Xfemales_converge[,6])
es_DGLM_Johansson_Blood_Xfemales<-es(bc_DGLM_Johansson_Blood_Xfemales_es)

# Summary Bacon statistics
dat_DGLM_Johansson_Blood_Xfemales_bacon_dispersion<-as.data.frame(cbind(es_DGLM_Johansson_Blood_Xfemales,tstats_DGLM_Johansson_Blood_Xfemales,pvals_DGLM_Johansson_Blood_Xfemales))
colnames(dat_DGLM_Johansson_Blood_Xfemales_bacon_dispersion)<-c("bacon_effectsize_dispersion","bacon_tstatistics_dispersion","bacon_pval_dispersion")
rownames(dat_DGLM_Johansson_Blood_Xfemales_bacon_dispersion)<-make.names(rownames(DGLM_Johansson_Blood_Xfemales_converge),unique = TRUE)

# check females effecsize of aVMCs between BIOS and Johansson 
dat_Johansson_Blood_Xfemales_aVMCs<-dat_DGLM_Johansson_Blood_Xfemales_bacon_dispersion[rownames(df_aVMCs_effectsize_BIOS),]
df_Johansson_validation_females<-cbind(dat_Johansson_Blood_Xfemales_aVMCs[,1],df_aVMCs_effectsize_BIOS[,-2])
colnames(df_Johansson_validation_females)<-c("Johansson","BIOS","symbol")

tiff("Output_06_Johansson_Validation of female effect size.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Johansson_validation_females) +
    aes(x = BIOS,y = Johansson,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Johansson Blood")
dev.off()

# Xmales
set.seed(1)
bc_DGLM_Johansson_Blood_Xmales<- bacon(DGLM_Johansson_Blood_Xmales_converge[,7])

# Calculate bias and inflation
bias(bc_DGLM_Johansson_Blood_Xmales)         # 0.40 
inflation(bc_DGLM_Johansson_Blood_Xmales)    # 0.93

# Save histogram plot of test statistics 
tiff("Output_06_Bacon_DGLM_Johansson_Blood_Xmales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Johansson_Blood_Xmales,xlab="t-statistics",main="aVMCs in Johansson Blood males",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Johansson_Blood_Xmales<-tstat(bc_DGLM_Johansson_Blood_Xmales)
pvals_DGLM_Johansson_Blood_Xmales <-pval(bc_DGLM_Johansson_Blood_Xmales)
table(pvals_DGLM_Johansson_Blood_Xmales<0.05)   # 998  
rownames(pvals_DGLM_Johansson_Blood_Xmales)<-make.names(rownames(DGLM_Johansson_Blood_Xmales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Johansson_Blood_Xmales_es<-bacon(NULL,DGLM_Johansson_Blood_Xmales_converge[,5],DGLM_Johansson_Blood_Xmales_converge[,6])
es_DGLM_Johansson_Blood_Xmales<-es(bc_DGLM_Johansson_Blood_Xmales_es)

# Summary Bacon statistics
dat_DGLM_Johansson_Blood_Xmales_bacon_dispersion<-as.data.frame(cbind(es_DGLM_Johansson_Blood_Xmales,tstats_DGLM_Johansson_Blood_Xmales,pvals_DGLM_Johansson_Blood_Xmales))
colnames(dat_DGLM_Johansson_Blood_Xmales_bacon_dispersion)<-c("bacon_effectsize_dispersion","bacon_tstatistics_dispersion","bacon_pval_dispersion")
rownames(dat_DGLM_Johansson_Blood_Xmales_bacon_dispersion)<-make.names(rownames(DGLM_Johansson_Blood_Xmales_converge),unique = TRUE)

# check males effecsize of aVMCs between BIOS and Johansson 
dat_Johansson_Blood_Xmales_aVMCs<-dat_DGLM_Johansson_Blood_Xmales_bacon_dispersion[rownames(df_aVMCs_effectsize_BIOS),]
df_Johansson_validation_males<-cbind(dat_Johansson_Blood_Xmales_aVMCs[,1],df_aVMCs_effectsize_BIOS[,-1])
colnames(df_Johansson_validation_males)<-c("Johansson","BIOS","symbol")

tiff("Output_06_Johansson_Validation of male effect size.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Johansson_validation_males) +
    aes(x = BIOS,y = Johansson,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Johansson Blood")
dev.off()
```

```{r}
# save the validation data of effect size 
save(df_Johansson_validation_females,df_Johansson_validation_males,file="Output_06_Johansson_validation.RData")
```

```{r}
# Extract 631 female p-values and 36 male p-values.
dat_Johansson_Blood_Xfemales_aVMCs$symbol<-df_aVMCs_effectsize_BIOS$symbol
df_aVMCs_females_Johansson_Blood<-filter(dat_Johansson_Blood_Xfemales_aVMCs,symbol=="female-specific")
pvals_aVMCs_Xfemales_Johansson_Blood<-as.data.frame(pvals_DGLM_Johansson_Blood_Xfemales[rownames(df_aVMCs_females_Johansson_Blood),])
colnames(pvals_aVMCs_Xfemales_Johansson_Blood)<-"pval_females"

dat_Johansson_Blood_Xmales_aVMCs$symbol<-df_aVMCs_effectsize_BIOS$symbol
df_aVMCs_males_Johansson_Blood<-filter(dat_Johansson_Blood_Xmales_aVMCs,symbol=="male-specific")
pvals_aVMCs_Xmales_Johansson_Blood<-as.data.frame(pvals_DGLM_Johansson_Blood_Xmales[rownames(df_aVMCs_males_Johansson_Blood),])
colnames(pvals_aVMCs_Xmales_Johansson_Blood)<-"pval_males"

# Perform FDR correction for 667 pvals(631+36)
padjs_aVMCs_Johansson_Blood<-as.data.frame(p.adjust(c(pvals_aVMCs_Xfemales_Johansson_Blood$pval_females,pvals_aVMCs_Xmales_Johansson_Blood$pval_males),method="BH"))
colnames(padjs_aVMCs_Johansson_Blood)<-"bacon_p.adj_dispersion"

# check females (how many of 631) and males (how many of 36) separately how many were FDR significant 
# 584 female aVMCs are still FDR significant in Hannum blood data.
padjs_aVMCs_females_Johansson_Blood<-as.data.frame(padjs_aVMCs_Johansson_Blood[1:631,])
colnames(padjs_aVMCs_females_Johansson_Blood)<-"bacon_p.adj_dispersion"
df_aVMCs_females_Johansson_Blood$bacon_p.adj_dispersion<-padjs_aVMCs_females_Johansson_Blood$bacon_p.adj_dispersion

df_aVMCs_females_Johansson_Blood<-filter(df_aVMCs_females_Johansson_Blood,bacon_p.adj_dispersion<0.05)
df_aVMCs_females_effectsize_up_Johansson<-filter(df_aVMCs_females_Johansson_Blood,bacon_effectsize_dispersion>0) # 582
df_aVMCs_females_effectsize_down_Johansson<-filter(df_aVMCs_females_Johansson_Blood,bacon_effectsize_dispersion<0) # 2

# 35 male aVMCs are still FDR significant in Hannum blood data.
padjs_aVMCs_males_Johansson_Blood<-as.data.frame(padjs_aVMCs_Johansson_Blood[632:667,])
colnames(padjs_aVMCs_males_Johansson_Blood)<-"bacon_p.adj_dispersion"
df_aVMCs_males_Johansson_Blood$bacon_p.adj_dispersion<-padjs_aVMCs_males_Johansson_Blood$bacon_p.adj_dispersion

df_aVMCs_males_Johansson_Blood<-filter(df_aVMCs_males_Johansson_Blood,bacon_p.adj_dispersion<0.05)
df_aVMCs_males_effectsize_up_Johansson<-filter(df_aVMCs_males_Johansson_Blood,bacon_effectsize_dispersion>0) # 34
df_aVMCs_males_effectsize_down_Johansson<-filter(df_aVMCs_males_Johansson_Blood,bacon_effectsize_dispersion<0) # 1

save(df_aVMCs_females_effectsize_up_Johansson,df_aVMCs_females_effectsize_down_Johansson,df_aVMCs_females_Johansson_Blood,df_aVMCs_males_effectsize_up_Johansson,df_aVMCs_males_effectsize_down_Johansson,df_aVMCs_males_Johansson_Blood,file = "Output_06_Johansson_Catogue of aVMCs.RData")
```

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```{r}
#### aVMCs validation on Liu Monocytes datasets
## Correct bias and inflation for dispersion model.
# Xfemales
set.seed(1)
bc_DGLM_Liu_Mono_Xfemales<- bacon(DGLM_Liu_Mono_Xfemales_converge[,7])

# Calculate bias and inflation
bias(bc_DGLM_Liu_Mono_Xfemales)         # 0.18  
inflation(bc_DGLM_Liu_Mono_Xfemales)    # 0.73

# Save histogram plot of test statistics 
tiff("Output_06_Bacon_DGLM_Liu_Mono_Xfemales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Liu_Mono_Xfemales,xlab="t-statistics",main="aVMCs in Liu Monocytes females",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Liu_Mono_Xfemales<-tstat(bc_DGLM_Liu_Mono_Xfemales)
pvals_DGLM_Liu_Mono_Xfemales <-pval(bc_DGLM_Liu_Mono_Xfemales)
table(pvals_DGLM_Liu_Mono_Xfemales<0.05)   # 6029  
rownames(pvals_DGLM_Liu_Mono_Xfemales)<-make.names(rownames(DGLM_Liu_Mono_Xfemales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Liu_Mono_Xfemales_es<-bacon(NULL,DGLM_Liu_Mono_Xfemales_converge[,5],DGLM_Liu_Mono_Xfemales_converge[,6])
es_DGLM_Liu_Mono_Xfemales<-es(bc_DGLM_Liu_Mono_Xfemales_es)

# Summary Bacon statistics
dat_DGLM_Liu_Mono_Xfemales_bacon_dispersion<-as.data.frame(cbind(es_DGLM_Liu_Mono_Xfemales,tstats_DGLM_Liu_Mono_Xfemales,pvals_DGLM_Liu_Mono_Xfemales))
colnames(dat_DGLM_Liu_Mono_Xfemales_bacon_dispersion)<-c("bacon_effectsize_dispersion","bacon_tstatistics_dispersion","bacon_pval_dispersion")
rownames(dat_DGLM_Liu_Mono_Xfemales_bacon_dispersion)<-make.names(rownames(DGLM_Liu_Mono_Xfemales_converge),unique = TRUE)

# check females effecsize of aVMCs between BIOS and Liu 
dat_Liu_Mono_Xfemales_aVMCs<-dat_DGLM_Liu_Mono_Xfemales_bacon_dispersion[rownames(df_aVMCs_effectsize_BIOS),]
df_Liu_validation_females<-cbind(dat_Liu_Mono_Xfemales_aVMCs[,1],df_aVMCs_effectsize_BIOS[,-2])
colnames(df_Liu_validation_females)<-c("Liu","BIOS","symbol")

tiff("Output_06_Liu_Validation of female effect size.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Liu_validation_females) +
    aes(x = BIOS,y = Liu,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Liu Monocytes")
dev.off()

# Xmales
set.seed(1)
bc_DGLM_Liu_Mono_Xmales<- bacon(DGLM_Liu_Mono_Xmales_converge[,7])

# Calculate bias and inflation
bias(bc_DGLM_Liu_Mono_Xmales)         # 0.31  
inflation(bc_DGLM_Liu_Mono_Xmales)    # 0.80

# Save histogram plot of test statistics 
tiff("Output_06_Bacon_DGLM_Liu_Mono_Xmales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Liu_Mono_Xmales,xlab="t-statistics",main="aVMCs in Liu Monocytes males",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Liu_Mono_Xmales<-tstat(bc_DGLM_Liu_Mono_Xmales)
pvals_DGLM_Liu_Mono_Xmales <-pval(bc_DGLM_Liu_Mono_Xmales)
table(pvals_DGLM_Liu_Mono_Xmales<0.05)   # 1389  
rownames(pvals_DGLM_Liu_Mono_Xmales)<-make.names(rownames(DGLM_Liu_Mono_Xmales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Liu_Mono_Xmales_es<-bacon(NULL,DGLM_Liu_Mono_Xmales_converge[,5],DGLM_Liu_Mono_Xmales_converge[,6])
es_DGLM_Liu_Mono_Xmales<-es(bc_DGLM_Liu_Mono_Xmales_es)

# Summary Bacon statistics
dat_DGLM_Liu_Mono_Xmales_bacon_dispersion<-as.data.frame(cbind(es_DGLM_Liu_Mono_Xmales,tstats_DGLM_Liu_Mono_Xmales,pvals_DGLM_Liu_Mono_Xmales))
colnames(dat_DGLM_Liu_Mono_Xmales_bacon_dispersion)<-c("bacon_effectsize_dispersion","bacon_tstatistics_dispersion","bacon_pval_dispersion")
rownames(dat_DGLM_Liu_Mono_Xmales_bacon_dispersion)<-make.names(rownames(DGLM_Liu_Mono_Xmales_converge),unique = TRUE)

# check males effecsize of aVMCs between BIOS and Liu 
dat_Liu_Mono_Xmales_aVMCs<-dat_DGLM_Liu_Mono_Xmales_bacon_dispersion[rownames(df_aVMCs_effectsize_BIOS),]
df_Liu_validation_males<-cbind(dat_Liu_Mono_Xmales_aVMCs[,1],df_aVMCs_effectsize_BIOS[,-1])
colnames(df_Liu_validation_males)<-c("Liu","BIOS","symbol")

tiff("Output_06_Liu_Validation of male effect size.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Liu_validation_males) +
    aes(x = BIOS,y = Liu,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Liu Monocytes")
dev.off()
```

```{r}
# save the validation data of effect size 
save(df_Liu_validation_females,df_Liu_validation_males,file="Output_06_Liu_validation.RData")
```

```{r}
# Extract 631 female p-values and 36 male p-values.
dat_Liu_Mono_Xfemales_aVMCs$symbol<-df_aVMCs_effectsize_BIOS$symbol
df_aVMCs_females_Liu_Mono<-filter(dat_Liu_Mono_Xfemales_aVMCs,symbol=="female-specific")
pvals_aVMCs_Xfemales_Liu_Mono<-as.data.frame(pvals_DGLM_Liu_Mono_Xfemales[rownames(df_aVMCs_females_Liu_Mono),])
colnames(pvals_aVMCs_Xfemales_Liu_Mono)<-"pval_females"

dat_Liu_Mono_Xmales_aVMCs$symbol<-df_aVMCs_effectsize_BIOS$symbol
df_aVMCs_males_Liu_Mono<-filter(dat_Liu_Mono_Xmales_aVMCs,symbol=="male-specific")
pvals_aVMCs_Xmales_Liu_Mono<-as.data.frame(pvals_DGLM_Liu_Mono_Xmales[rownames(df_aVMCs_males_Liu_Mono),])
colnames(pvals_aVMCs_Xmales_Liu_Mono)<-"pval_males"

# Perform FDR correction for 667 pvals(631+36)
padjs_aVMCs_Liu_Mono<-as.data.frame(p.adjust(c(pvals_aVMCs_Xfemales_Liu_Mono$pval_females,pvals_aVMCs_Xmales_Liu_Mono$pval_males),method="BH"))
colnames(padjs_aVMCs_Liu_Mono)<-"bacon_p.adj_dispersion"

# check females (how many of 631) and males (how many of 36) separately how many were FDR significant 
# 597 female aVMCs are still FDR significant in Liu Monocytes data.
padjs_aVMCs_females_Liu_Mono<-as.data.frame(padjs_aVMCs_Liu_Mono[1:631,])
colnames(padjs_aVMCs_females_Liu_Mono)<-"bacon_p.adj_dispersion"
df_aVMCs_females_Liu_Mono$bacon_p.adj_dispersion<-padjs_aVMCs_females_Liu_Mono$bacon_p.adj_dispersion

df_aVMCs_females_Liu_Mono<-filter(df_aVMCs_females_Liu_Mono,bacon_p.adj_dispersion<0.05)
df_aVMCs_females_effectsize_up_Liu<-filter(df_aVMCs_females_Liu_Mono,bacon_effectsize_dispersion>0) # 597
df_aVMCs_females_effectsize_down_Liu<-filter(df_aVMCs_females_Liu_Mono,bacon_effectsize_dispersion<0) # 0

# 34 male aVMCs are still FDR significant in Hannum blood data.
padjs_aVMCs_males_Liu_Mono<-as.data.frame(padjs_aVMCs_Liu_Mono[632:667,])
colnames(padjs_aVMCs_males_Liu_Mono)<-"bacon_p.adj_dispersion"
df_aVMCs_males_Liu_Mono$bacon_p.adj_dispersion<-padjs_aVMCs_males_Liu_Mono$bacon_p.adj_dispersion

df_aVMCs_males_Liu_Mono<-filter(df_aVMCs_males_Liu_Mono,bacon_p.adj_dispersion<0.05)
df_aVMCs_males_effectsize_up_Liu<-filter(df_aVMCs_males_Liu_Mono,bacon_effectsize_dispersion>0) # 33
df_aVMCs_males_effectsize_down_Liu<-filter(df_aVMCs_males_Liu_Mono,bacon_effectsize_dispersion<0) # 1

save(df_aVMCs_females_effectsize_up_Liu,df_aVMCs_females_Liu_Mono,df_aVMCs_males_effectsize_up_Liu,df_aVMCs_males_effectsize_down_Liu,df_aVMCs_males_Liu_Mono,file = "Output_06_Liu_Catogue of aVMCs.RData")
```

```{r}
# The catlogue of validated aDMCs in females and males.
# Females:532
Overlap_females<-Reduce(intersect, list(rownames(dat_DGLM_Xfemales_bacon_disp),rownames(df_aVMCs_females_effectsize_up_Johansson),rownames(df_aVMCs_females_Hannum_Blood),rownames(df_aVMCs_females_Liu_Mono)))
dat_aVMCs_females_validated<-dat_DGLM_Xfemales_bacon_disp[Overlap_females,]

# Males: 34
Overlap_males<-Reduce(intersect, list(rownames(dat_DGLM_Xmales_bacon_disp),rownames(df_aVMCs_males_Johansson_Blood),rownames(df_aVMCs_males_Hannum_Blood),rownames(df_aVMCs_males_Liu_Mono)))
dat_aVMCs_males_validated<-dat_DGLM_Xmales_bacon_disp[Overlap_males,]

save(dat_aVMCs_females_validated,dat_aVMCs_males_validated,file = "Output_06_Catalogue_aVMCs_validated.RData")
```

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```{r}
library(gridExtra)
tiff("Output_06_Validation of aVMCs.tiff", units="in", width=18, height=14, res=800,compression = 'lzw')
p1<-ggplot(df_Johansson_validation_females) +
    aes(x = BIOS,y = Johansson,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Johansson Blood")

p2<-ggplot(df_Johansson_validation_males) +
    aes(x = BIOS,y = Johansson,color=symbol) +
     geom_point(size = 2, alpha = 0.5) +
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Johansson Blood")
  
p3<-ggplot(df_Hannum_validation_females) +
    aes(x = BIOS,y = Hannum,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Hannum Blood")
  
p4<-ggplot(df_Hannum_validation_males) +
    aes(x = BIOS,y = Hannum,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Hannum Blood")
  
p5<-ggplot(df_Liu_validation_females) +
    aes(x = BIOS,y = Liu,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Liu Monocytes")
  
p6<-ggplot(df_Liu_validation_males) +
    aes(x = BIOS,y = Liu,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aVMCs",breaks=c("female-specific","male-specific"),labels=c("female-specific:631CpGs","male-specific:36CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Liu Monocytes")
grid.arrange(p1,p2,p3,p4,p5,p6,ncol=2)
dev.off()
```

```{r}
# The overlap aVMCs between BIOS data and three external datasets.
# Venn plot
library(RColorBrewer)
myCol <- brewer.pal(4,"Pastel2")
display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}

# The statistically significant overlapping aVMCs in females
df_aVMCs_females<-list(Hannum=rownames(df_aVMCs_females_Hannum_Blood), Johansson =rownames(df_aVMCs_females_effectsize_up_Johansson),Liu=rownames(df_aVMCs_females_Liu_Mono), BIOS=rownames(dat_DGLM_Xfemales_bacon_disp))

# The statistically significant overlapping aDMCs in males 
df_aVMCs_males<-list(Hannum=rownames(df_aVMCs_males_Hannum_Blood), Johansson =rownames(df_aVMCs_males_Johansson_Blood),Liu=rownames(df_aVMCs_males_Liu_Mono), BIOS=rownames(dat_DGLM_Xmales_bacon_disp))

# Venn plot.
tiff("Output_06_venn_aVMCs_females.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aVMCs_females,fill = myCol,cat.fontface = "bold",main="Females",main.fontface="bold",cex = 1.5,main.cex = 1.5,cat.cex=1.5)
dev.off()

tiff("Output_06_venn_aVMCs_males.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aVMCs_males,fill = myCol,cat.fontface = "bold",main="Males",main.fontface="bold",cex = 1.5,main.cex = 1.5,cat.cex=1.5)
dev.off()
```

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```{r}
# The overlap aDMCs between BIOS data and three external datasets.
# Venn plot
library(RColorBrewer)
myCol <- brewer.pal(4,"Pastel2")
display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}

# The statistically significant overlapping aVMCs in females (effectsize>0).
df_aVMCs_females_effectsize_up<-list(Hannum=rownames(df_aVMCs_females_effectsize_up_Hannum), Johansson =rownames(df_aVMCs_females_effectsize_up_Johansson),Liu=rownames(df_aVMCs_females_effectsize_up_Liu), BIOS=rownames(dat_DGLM_Xfemales_bacon_disp_up))

tiff("Output_06_venn_aVMCs_effectsize_up_females.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aVMCs_females_effectsize_up,fill = myCol,cat.fontface = "bold",main="Statistically significant overlapping aVMCs in females(effect size>0)",main.fontface="bold")
dev.off()

# The statistically significant overlapping aVMCs in males (effectsize<0).
df_aVMCs_males_effectsize_down<-list(Hannum=rownames(df_aVMCs_males_effectsize_down_Hannum), Johansson =rownames(df_aVMCs_males_effectsize_down_Johansson),Liu=rownames(df_aVMCs_males_effectsize_down_Liu), BIOS=rownames(dat_DGLM_Xmales_bacon_disp_down))

tiff("Output_06_venn_aVMCs_effectsize_down_males.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aVMCs_males_effectsize_down,fill = myCol,cat.fontface = "bold",main="Statistically significant overlapping aVMCs in males(effect size<0)",main.fontface="bold")
dev.off()

# The statistically significant overlapping aVMCs in males (effectsize>0).
df_aVMCs_males_effectsize_up<-list(Hannum=rownames(df_aVMCs_males_effectsize_up_Hannum), Johansson =rownames(df_aVMCs_males_effectsize_up_Johansson),Liu=rownames(df_aVMCs_males_effectsize_up_Liu), BIOS=rownames(dat_DGLM_Xmales_bacon_disp_up))

tiff("Output_06_venn_aVMCs_effectsize_up_males.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aVMCs_males_effectsize_up,fill = myCol,cat.fontface = "bold",main="Statistically significant overlapping aVMCs in males(effect size>0)",main.fontface="bold")
dev.off()
```











































































