---
title: "aDMCs Validation"
author: "Yunfeng Liu"
date: "5/18/2022"
output: html_document
---
Here,we would like to replicate our aDMCs analysis on X chromosome using three external datasets.

1.Blood DNAm data generated by Hannum et al.: "Genome-wide Methylation Profiles Reveal Quantitative Views of Human Aging Rates.Mol Cell."
GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40279

2.Blood DNAm data generated by Johansson et al.: "Continuous Aging of the Human DNA Methylome Throughout the Human Lifespan.PLoS One."
GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87571

3.Monocytes DNAm data generated by Liu et al.: "Age-related variations in the methylome associated with gene expression in human monocytes and T cells. Nat Commun."
GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE56046


```{r}
#set library path.
.libPaths("~/researchdrive/yliu/BIOS_all/Functional interpretation of age-related X methylation/Rlibs")
```


```{r}
#Load all necessary libraries.
library(minfi)
library(DNAmArray)
library(tidyverse)
library(sva)
library(bacon)
library(irlba)
library(gridExtra)
```


```{r}
# Load data
load(file="Input_05_Hannum_Blood_450k.RData")
load(file = "Input_05_Johansson_Blood_450k.RData")
load(file = "Input_05_Liu_Monocytes_450k.RData")
load(file = "Output_02_aDMCs_effectsize_BIOS.RData")
```


```{r}
#### aDMCs validation on Hannum blood datasets
# rank-inverse normal(RIN) transformation for Xfemales and Xmales separately.
RIN <- function(x) {
  y <- rank(x, NA)
  y <- ppoints(y)[y]
  y <- qnorm(y)
  x[!is.na(x)] <- y
  x
}

RIN.mvalues_Hannum_Blood_Xfemales <- t(apply(mvalues_Hannum_Blood_Xfemales, 1, RIN))
RIN.mvalues_Hannum_Blood_Xmales <- t(apply(mvalues_Hannum_Blood_Xmales, 1, RIN))
```


```{r}
# Create PCs of the Hannum data for Xfemales, Xmales.
pc_Hannum_Blood_Xfemales<- prcomp_irlba(t(RIN.mvalues_Hannum_Blood_Xfemales), n=10)
pc_Hannum_Blood_Xmales<- prcomp_irlba(t(RIN.mvalues_Hannum_Blood_Xmales), n=10)

# Screenplot drawing
summary(pc_Hannum_Blood_Xfemales)
png("Output_05_screenplot_Hannum_Blood_Xfemales.png")
screeplot(pc_Hannum_Blood_Xfemales,type = "lines",main = "Hannum_Blood_Xfemales")
dev.off()

summary(pc_Hannum_Blood_Xmales)
png("Output_05_screenplot_Hannum_Blood_Xmales.png")
screeplot(pc_Hannum_Blood_Xmales,type = "lines",main = "Hannum_Blood_Xmales")
dev.off()
```


```{r}
## Prepare SVA
# Hannum Xfemales
## null model:only includes the intercept
design0_Hannum_Blood_Xfemales = model.matrix(~ 1,data=Covariates_Hannum_Blood_Xfemales)
## Full model:all covarites
design_Hannum_Blood_Xfemales <- model.matrix(~Age + CD8T + CD4T + NK + Bcell + Mono + sentrix_position, data=Covariates_Hannum_Blood_Xfemales)

## Estimate Hannum female latent factors by SVA
svobj_Hannum_Blood_Xfemales = sva(RIN.mvalues_Hannum_Blood_Xfemales,design_Hannum_Blood_Xfemales,design0_Hannum_Blood_Xfemales,n.sv=5)
sv_Hannum_Blood_Xfemales <- as.data.frame(svobj_Hannum_Blood_Xfemales$sv)
colnames(sv_Hannum_Blood_Xfemales) <- c('SV1', 'SV2', 'SV3', 'SV4', 'SV5')

# Hannum Xmales
## null model:only includes the intercept
design0_Hannum_Blood_Xmales = model.matrix(~ 1,data=Covariates_Hannum_Blood_Xmales)
## Full model:all covarites
design_Hannum_Blood_Xmales <- model.matrix(~Age + CD8T + CD4T + NK + Bcell + Mono + sentrix_position, data=Covariates_Hannum_Blood_Xmales)

# Estimate Hannum male latent factors by SVA
svobj_Hannum_Blood_Xmales = sva(RIN.mvalues_Hannum_Blood_Xmales,design_Hannum_Blood_Xmales,design0_Hannum_Blood_Xmales,n.sv=5)
sv_Hannum_Blood_Xmales <- as.data.frame(svobj_Hannum_Blood_Xmales$sv)
colnames(sv_Hannum_Blood_Xmales) <- c('SV1', 'SV2', 'SV3', 'SV4', 'SV5')
```


```{r}
# Add estimated latent factors to covariates dataframe. 
#Xfemales
Covariates_sv_Hannum_Blood_Xfemales<-cbind(Covariates_Hannum_Blood_Xfemales,sv_Hannum_Blood_Xfemales)

#Xmales
Covariates_sv_Hannum_Blood_Xmales<-cbind(Covariates_Hannum_Blood_Xmales,sv_Hannum_Blood_Xmales)

# Change sentrixt position into factor
Covariates_sv_Hannum_Blood_Xfemales$sentrix_position<-factor(Covariates_sv_Hannum_Blood_Xfemales$sentrix_position)
Covariates_sv_Hannum_Blood_Xmales$sentrix_position<-factor(Covariates_sv_Hannum_Blood_Xmales$sentrix_position)
```

```{r}
#### Construct a Function called runDGLM for Xfemales and Xmales separately.
## Identify aDMCs in Hannum Blood females.
runDGLM_Hannum_Blood_Xfemales <- function(x)
{
  dat_Xfemales_dglm <- matrix(nrow=nrow(x), ncol=9)
  colnames(dat_Xfemales_dglm)<-c("cgID","effectsize.linear","std.error.linear","t.linear","p.linear","effectsize.disp","std.error.disp","t.disp","p.disp")
  for(CpG in 1:nrow(x))
  {
    print(CpG) #to check progress
    probe <- rownames(x)[CpG]
    tempdata <- data.frame(Age=Covariates_sv_Hannum_Blood_Xfemales[,3],data = x[probe,],sentrix_position=Covariates_sv_Hannum_Blood_Xfemales[,5],CD8T=Covariates_sv_Hannum_Blood_Xfemales[,6], CD4T=Covariates_sv_Hannum_Blood_Xfemales[,7], NK=Covariates_sv_Hannum_Blood_Xfemales[,8], Bcell=Covariates_sv_Hannum_Blood_Xfemales[,9], Mono=Covariates_sv_Hannum_Blood_Xfemales[,10],  SV1=Covariates_sv_Hannum_Blood_Xfemales[,12], SV2=Covariates_sv_Hannum_Blood_Xfemales[,13], SV3=Covariates_sv_Hannum_Blood_Xfemales[,14], SV4=Covariates_sv_Hannum_Blood_Xfemales[,15], SV5=Covariates_sv_Hannum_Blood_Xfemales[,16])
    
    fit_Xfemales <- tryCatch({ dglm(formula = data ~ Age + CD8T + CD4T + NK + Bcell + Mono + sentrix_position + SV1 + SV2 + SV3 + SV4 + SV5, dformula = ~ Age, data=tempdata, family = gaussian(link = identity))},error=identity)
    if(is.null(fit_Xfemales$message)){ #good
      #extract info about the linear model
      effectsize.linear <- summary(fit_Xfemales)$coefficients[2,1]
      std.error.linear<-summary(fit_Xfemales)$coefficients[2,2]
      t.value.linear <-summary(fit_Xfemales)$coefficients[2,3]
      p.value.linear <- summary(fit_Xfemales)$coefficients[2,4]
      
      #extract info about the dispersion model
      effectsize.disp <- summary(fit_Xfemales$dispersion.fit)$coefficients[2]
      std.error.disp<-summary(fit_Xfemales$dispersion.fit)$coefficients[4]
      t.value.disp <- summary(fit_Xfemales$dispersion.fit)$coefficients[6]
      p.value.disp <- summary(fit_Xfemales$dispersion.fit)$coefficients[8]
      
      # Summary statistics Output
      out <- matrix(c(probe,effectsize.linear,std.error.linear,t.value.linear,p.value.linear,effectsize.disp,std.error.disp,t.value.disp,p.value.disp), ncol=9)
      dat_Xfemales_dglm[CpG,] <- out
    }else{                  #bad
      print("Didn't converge")
      out <- matrix(c(probe, fit_Xfemales$message, rep(NA, times = 7)), ncol=9)
      dat_Xfemales_dglm[CpG,] <- out
    }
    
  }
  return(dat_Xfemales_dglm)
}

# Run runDGLM function
DGLM_Hannum_Blood_Xfemales<- runDGLM_Hannum_Blood_Xfemales(RIN.mvalues_Hannum_Blood_Xfemales)
DGLM_Hannum_Blood_Xfemales<-data.frame(DGLM_Hannum_Blood_Xfemales) 
rownames(DGLM_Hannum_Blood_Xfemales)<-make.names(DGLM_Hannum_Blood_Xfemales$cgID,unique=TRUE)
DGLM_Hannum_Blood_Xfemales<-DGLM_Hannum_Blood_Xfemales[,2:9]
DGLM_Hannum_Blood_Xfemales_errors<-subset(DGLM_Hannum_Blood_Xfemales,is.na(DGLM_Hannum_Blood_Xfemales$t.linear)) # No errors(all CpGs converge)
DGLM_Hannum_Blood_Xfemales_converge<-DGLM_Hannum_Blood_Xfemales

# Convert each column to numerical variable.
DGLM_Hannum_Blood_Xfemales_converge<-as.data.frame(lapply(DGLM_Hannum_Blood_Xfemales_converge,as.numeric)) 
rownames(DGLM_Hannum_Blood_Xfemales_converge)<-make.names(rownames(DGLM_Hannum_Blood_Xfemales),unique=TRUE) 
save(DGLM_Hannum_Blood_Xfemales_converge,file = "Output_05_DGLM_Hannum_Blood_Xfemales_converge.RData")
```

```{r}
## Correct bias and inflation for linear model.
# Xfemales
set.seed(1)
bc_DGLM_Hannum_Blood_Xfemales<- bacon(DGLM_Hannum_Blood_Xfemales_converge[,3])

# Calculate bias and inflation
bias(bc_DGLM_Hannum_Blood_Xfemales)         # -0.02  
inflation(bc_DGLM_Hannum_Blood_Xfemales)    # 1.3

# Save histogram plot of test statistics 
tiff("Output_05_Bacon_DGLM_Hannum_Blood_Xfemales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Hannum_Blood_Xfemales,xlab="t-statistics",main="aDMCs in Hannum Blood females",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Hannum_Blood_Xfemales<-tstat(bc_DGLM_Hannum_Blood_Xfemales)
pvals_DGLM_Hannum_Blood_Xfemales <-pval(bc_DGLM_Hannum_Blood_Xfemales)
table(pvals_DGLM_Hannum_Blood_Xfemales<0.05)   # 729  
rownames(pvals_DGLM_Hannum_Blood_Xfemales)<-make.names(rownames(DGLM_Hannum_Blood_Xfemales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Hannum_Blood_Xfemales_es<-bacon(NULL,DGLM_Hannum_Blood_Xfemales_converge[,1],DGLM_Hannum_Blood_Xfemales_converge[,2])
es_DGLM_Hannum_Blood_Xfemales<-es(bc_DGLM_Hannum_Blood_Xfemales_es)

# Summary Bacon statistics
dat_DGLM_Hannum_Blood_Xfemales_bacon_linear<-as.data.frame(cbind(es_DGLM_Hannum_Blood_Xfemales,tstats_DGLM_Hannum_Blood_Xfemales,pvals_DGLM_Hannum_Blood_Xfemales))
colnames(dat_DGLM_Hannum_Blood_Xfemales_bacon_linear)<-c("bacon_effectsize_linear","bacon_tstatistics_linear","bacon_pval_linear")
rownames(dat_DGLM_Hannum_Blood_Xfemales_bacon_linear)<-make.names(rownames(DGLM_Hannum_Blood_Xfemales_converge),unique = TRUE)

# check females effecsize of aDMCs between BIOS and Hannum 
dat_Hannum_Blood_Xfemales_aDMCs<-dat_DGLM_Hannum_Blood_Xfemales_bacon_linear[rownames(df_aDMCs_effectsize_BIOS),]
df_Hannum_validation_females<-cbind(dat_Hannum_Blood_Xfemales_aDMCs[,1],df_aDMCs_effectsize_BIOS[,-2])
colnames(df_Hannum_validation_females)<-c("Hannum","BIOS","symbol")

tiff("Output_05_Hannum_Validation of female size effect.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Hannum_validation_females) +
    aes(x = BIOS,y = Hannum,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Hannum Blood")
dev.off()
```


```{r}
#### Construct a Function called runDGLM for Xfemales and Xmales separately.
## Identify aDMCs in Hannum Blood males.
runDGLM_Hannum_Blood_Xmales <- function(x)
{
  dat_Xmales_dglm <- matrix(nrow=nrow(x), ncol=9)
  colnames(dat_Xmales_dglm)<-c("cgID","effectsize.linear","std.error.linear","t.linear","p.linear","effectsize.disp","std.error.disp","t.disp","p.disp")
  for(CpG in 1:nrow(x))
  {
    print(CpG) #to check progress
    probe <- rownames(x)[CpG]
    tempdata <- data.frame(Age=Covariates_sv_Hannum_Blood_Xmales[,3],data = x[probe,],sentrix_position=Covariates_sv_Hannum_Blood_Xmales[,5],CD8T=Covariates_sv_Hannum_Blood_Xmales[,6], CD4T=Covariates_sv_Hannum_Blood_Xmales[,7], NK=Covariates_sv_Hannum_Blood_Xmales[,8], Bcell=Covariates_sv_Hannum_Blood_Xmales[,9], Mono=Covariates_sv_Hannum_Blood_Xmales[,10],  SV1=Covariates_sv_Hannum_Blood_Xmales[,12], SV2=Covariates_sv_Hannum_Blood_Xmales[,13], SV3=Covariates_sv_Hannum_Blood_Xmales[,14], SV4=Covariates_sv_Hannum_Blood_Xmales[,15], SV5=Covariates_sv_Hannum_Blood_Xmales[,16])
    
    fit_Xmales <- tryCatch({ dglm(formula = data ~ Age + CD8T + CD4T + NK + Bcell + Mono + sentrix_position + SV1 + SV2 + SV3 + SV4 + SV5, dformula = ~ Age, data=tempdata, family = gaussian(link = identity))},error=identity)
    if(is.null(fit_Xmales$message)){ #good
      #extract info about the linear model
      effectsize.linear <- summary(fit_Xmales)$coefficients[2,1]
      std.error.linear<-summary(fit_Xmales)$coefficients[2,2]
      t.value.linear <-summary(fit_Xmales)$coefficients[2,3]
      p.value.linear <- summary(fit_Xmales)$coefficients[2,4]
      
      #extract info about the dispersion model
      effectsize.disp <- summary(fit_Xmales$dispersion.fit)$coefficients[2]
      std.error.disp<-summary(fit_Xmales$dispersion.fit)$coefficients[4]
      t.value.disp <- summary(fit_Xmales$dispersion.fit)$coefficients[6]
      p.value.disp <- summary(fit_Xmales$dispersion.fit)$coefficients[8]
      
      # Summary statistics Output
      out <- matrix(c(probe,effectsize.linear,std.error.linear,t.value.linear,p.value.linear,effectsize.disp,std.error.disp,t.value.disp,p.value.disp), ncol=9)
      dat_Xmales_dglm[CpG,] <- out
    }else{                  #bad
      print("Didn't converge")
      out <- matrix(c(probe, fit_Xmales$message, rep(NA, times = 7)), ncol=9)
      dat_Xmales_dglm[CpG,] <- out
    }
    
  }
  return(dat_Xmales_dglm)
}

# Run runDGLM function
DGLM_Hannum_Blood_Xmales<- runDGLM_Hannum_Blood_Xmales(RIN.mvalues_Hannum_Blood_Xmales)
DGLM_Hannum_Blood_Xmales<-data.frame(DGLM_Hannum_Blood_Xmales) 
rownames(DGLM_Hannum_Blood_Xmales)<-make.names(DGLM_Hannum_Blood_Xmales$cgID,unique=TRUE)
DGLM_Hannum_Blood_Xmales<-DGLM_Hannum_Blood_Xmales[,2:9]
DGLM_Hannum_Blood_Xmales_errors<-subset(DGLM_Hannum_Blood_Xmales,is.na(DGLM_Hannum_Blood_Xmales$t.linear)) # No errors(all CpGs converge)
DGLM_Hannum_Blood_Xmales_converge<-DGLM_Hannum_Blood_Xmales

# Convert each column to numerical variable.
DGLM_Hannum_Blood_Xmales_converge<-as.data.frame(lapply(DGLM_Hannum_Blood_Xmales_converge,as.numeric)) 
rownames(DGLM_Hannum_Blood_Xmales_converge)<-make.names(rownames(DGLM_Hannum_Blood_Xmales),unique=TRUE)
save(DGLM_Hannum_Blood_Xmales_converge,file = "Output_05_DGLM_Hannum_Blood_Xmales_converge.RData")
```


```{r}
## Correct bias and inflation for linear model.
# Xmales
set.seed(1)
bc_DGLM_Hannum_Blood_Xmales<- bacon(DGLM_Hannum_Blood_Xmales_converge[,3])

# Calculate bias and inflation
bias(bc_DGLM_Hannum_Blood_Xmales)         # -0.08  
inflation(bc_DGLM_Hannum_Blood_Xmales)    # 1.7

# Save histogram plot of test statistics 
tiff("Output_05_Bacon_DGLM_Hannum_Blood_Xmales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Hannum_Blood_Xmales,xlab="t-statistics",main="aDMCs in Hannum Blood males",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Hannum_Blood_Xmales<-tstat(bc_DGLM_Hannum_Blood_Xmales)
pvals_DGLM_Hannum_Blood_Xmales <-pval(bc_DGLM_Hannum_Blood_Xmales)
table(pvals_DGLM_Hannum_Blood_Xmales<0.05)   # 1163 
rownames(pvals_DGLM_Hannum_Blood_Xmales)<-make.names(rownames(DGLM_Hannum_Blood_Xmales_converge),unique=TRUE)

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Hannum_Blood_Xmales_es<-bacon(NULL,DGLM_Hannum_Blood_Xmales_converge[,1],DGLM_Hannum_Blood_Xmales_converge[,2])
es_DGLM_Hannum_Blood_Xmales<-es(bc_DGLM_Hannum_Blood_Xmales_es)

# Summary Bacon statistics
dat_DGLM_Hannum_Blood_Xmales_bacon_linear<-as.data.frame(cbind(es_DGLM_Hannum_Blood_Xmales,tstats_DGLM_Hannum_Blood_Xmales,pvals_DGLM_Hannum_Blood_Xmales))
colnames(dat_DGLM_Hannum_Blood_Xmales_bacon_linear)<-c("bacon_effectsize_linear","bacon_tstatistics_linear","bacon_pval_linear")
rownames(dat_DGLM_Hannum_Blood_Xmales_bacon_linear)<-make.names(rownames(DGLM_Hannum_Blood_Xmales_converge),unique = TRUE)

# check males effecsize of aDMCs between BIOS and Hannum 
dat_Hannum_Blood_Xmales_aDMCs<-dat_DGLM_Hannum_Blood_Xmales_bacon_linear[rownames(df_aDMCs_effectsize_BIOS),]
df_Hannum_validation_males<-cbind(dat_Hannum_Blood_Xmales_aDMCs[,1],df_aDMCs_effectsize_BIOS[,-1])
colnames(df_Hannum_validation_males)<-c("Hannum","BIOS","symbol")

tiff("Output_05_Hannum_Validation of male size effect.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Hannum_validation_males) +
    aes(x = BIOS,y = Hannum,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effectsize in BIOS",y="effectsize in Hannum Blood")
dev.off()
```

```{r}
# save the validation data of effect size 
save(df_Hannum_validation_females,df_Hannum_validation_males,file="Output_05_Hannum_validation.RData")
```


```{r}
# Extract 66(34+32) female p-values and 1522(1490+32) male p-values.
dat_Hannum_Blood_Xfemales_aDMCs$symbol<-df_aDMCs_effectsize_BIOS$symbol
df_aDMCs_females_Hannum_Blood<-filter(dat_Hannum_Blood_Xfemales_aDMCs,symbol=="female-specific"|symbol=="sex-independent")
pvals_aDMCs_Xfemales_Hannum_Blood<-as.data.frame(pvals_DGLM_Hannum_Blood_Xfemales[rownames(df_aDMCs_females_Hannum_Blood),])
colnames(pvals_aDMCs_Xfemales_Hannum_Blood)<-"pval_females"

dat_Hannum_Blood_Xmales_aDMCs$symbol<-df_aDMCs_effectsize_BIOS$symbol
df_aDMCs_males_Hannum_Blood<-filter(dat_Hannum_Blood_Xmales_aDMCs,symbol=="male-specific"|symbol=="sex-independent")
pvals_aDMCs_Xmales_Hannum_Blood<-as.data.frame(pvals_DGLM_Hannum_Blood_Xmales[rownames(df_aDMCs_males_Hannum_Blood),])
colnames(pvals_aDMCs_Xmales_Hannum_Blood)<-"pval_males"

# Perform FDR correction for 1588 pvals(66+1522)
padjs_aDMCs_Hannum_Blood<-as.data.frame(p.adjust(c(pvals_aDMCs_Xfemales_Hannum_Blood$pval_females,pvals_aDMCs_Xmales_Hannum_Blood$pval_males),method="BH"))
colnames(padjs_aDMCs_Hannum_Blood)<-"bacon_p.adj_linear"

# check males (how many of 1522) and females (how many of 66) separately how many were FDR significant 
# 53 female aDMCs are still FDR significant in Hannum blood data.
padjs_aDMCs_females_Hannum_Blood<-as.data.frame(padjs_aDMCs_Hannum_Blood[1:66,])
colnames(padjs_aDMCs_females_Hannum_Blood)<-"bacon_p.adj_linear"
df_aDMCs_females_Hannum_Blood$bacon_p.adj_linear<-padjs_aDMCs_females_Hannum_Blood$bacon_p.adj_linear

df_aDMCs_females_Hannum_Blood<-filter(df_aDMCs_females_Hannum_Blood,bacon_p.adj_linear<0.05)
df_aDMCs_females_effectsize_up_Hannum<-filter(df_aDMCs_females_Hannum_Blood,bacon_effectsize_linear>0) # 27
df_aDMCs_females_effectsize_down_Hannum<-filter(df_aDMCs_females_Hannum_Blood,bacon_effectsize_linear<0) # 26

# 499 male aDMCs are still FDR significant in Hannum blood data.
padjs_aDMCs_males_Hannum_Blood<-as.data.frame(padjs_aDMCs_Hannum_Blood[67:1588,])
colnames(padjs_aDMCs_males_Hannum_Blood)<-"bacon_p.adj_linear"
df_aDMCs_males_Hannum_Blood$bacon_p.adj_linear<-padjs_aDMCs_males_Hannum_Blood$bacon_p.adj_linear

df_aDMCs_males_Hannum_Blood<-filter(df_aDMCs_males_Hannum_Blood,bacon_p.adj_linear<0.05)
df_aDMCs_males_effectsize_up_Hannum<-filter(df_aDMCs_males_Hannum_Blood,bacon_effectsize_linear>0) # 310
df_aDMCs_males_effectsize_down_Hannum<-filter(df_aDMCs_males_Hannum_Blood,bacon_effectsize_linear<0) # 189

save(df_aDMCs_females_effectsize_up_Hannum,df_aDMCs_females_effectsize_down_Hannum,df_aDMCs_females_Hannum_Blood,df_aDMCs_males_effectsize_up_Hannum,df_aDMCs_males_effectsize_down_Hannum,df_aDMCs_males_Hannum_Blood,file = "Output_05_Hannum_Catogue of aDMCs.RData")
```

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```{r}
#### aDMCs validation on Johansson blood datasets.
# rank-inverse normal(RIN) transformation for Xfemales and Xmales separately.
RIN.mvalues_Johansson_Blood_Xfemales <- t(apply(mvalues_Johansson_Blood_Xfemales, 1, RIN))
RIN.mvalues_Johansson_Blood_Xmales <- t(apply(mvalues_Johansson_Blood_Xmales, 1, RIN))
```

```{r}
# Create PCs of the Johansson data for Xfemales, Xmales.
pc_Johansson_Blood_Xfemales<- prcomp_irlba(t(RIN.mvalues_Johansson_Blood_Xfemales), n=10)
pc_Johansson_Blood_Xmales<- prcomp_irlba(t(RIN.mvalues_Johansson_Blood_Xmales), n=10)

# Screenplot drawing
summary(pc_Johansson_Blood_Xfemales)
png("Output_05_screenplot_Johansson_Blood_Xfemales.png")
screeplot(pc_Johansson_Blood_Xfemales,type = "lines",main = "Johansson_Blood_Xfemales")
dev.off()

summary(pc_Johansson_Blood_Xmales)
png("Output_05_screenplot_Johansson_Blood_Xmales.png")
screeplot(pc_Johansson_Blood_Xmales,type = "lines",main = "Johansson_Blood_Xmales")
dev.off()
```

```{r}
## Prepare SVA
# Johansson Xfemales
## null model:only includes the intercept
design0_Johansson_Blood_Xfemales = model.matrix(~ 1,data=Covariates_Johansson_Blood_Xfemales)
## Full model:all covarites
design_Johansson_Blood_Xfemales <- model.matrix(~Age + CD8T + CD4T + NK + Bcell + Mono + sentrix_position, data=Covariates_Johansson_Blood_Xfemales)

## Estimate Johansson female latent factors by SVA
svobj_Johansson_Blood_Xfemales = sva(RIN.mvalues_Johansson_Blood_Xfemales,design_Johansson_Blood_Xfemales,design0_Johansson_Blood_Xfemales,n.sv=5)
sv_Johansson_Blood_Xfemales <- as.data.frame(svobj_Johansson_Blood_Xfemales$sv)
colnames(sv_Johansson_Blood_Xfemales) <- c('SV1', 'SV2', 'SV3', 'SV4', 'SV5')

# Johansson Xmales
## null model:only includes the intercept
design0_Johansson_Blood_Xmales = model.matrix(~ 1,data=Covariates_Johansson_Blood_Xmales)
## Full model:all covarites
design_Johansson_Blood_Xmales <- model.matrix(~Age + CD8T + CD4T + NK + Bcell + Mono + sentrix_position, data=Covariates_Johansson_Blood_Xmales)

# Estimate Johansson male latent factors by SVA
svobj_Johansson_Blood_Xmales = sva(RIN.mvalues_Johansson_Blood_Xmales,design_Johansson_Blood_Xmales,design0_Johansson_Blood_Xmales,n.sv=5)
sv_Johansson_Blood_Xmales <- as.data.frame(svobj_Johansson_Blood_Xmales$sv)
colnames(sv_Johansson_Blood_Xmales) <- c('SV1', 'SV2', 'SV3', 'SV4', 'SV5')
```

```{r}
# Add estimated latent factors to covariates dataframe. 
#Xfemales
Covariates_sv_Johansson_Blood_Xfemales<-cbind(Covariates_Johansson_Blood_Xfemales,sv_Johansson_Blood_Xfemales)

#Xmales
Covariates_sv_Johansson_Blood_Xmales<-cbind(Covariates_Johansson_Blood_Xmales,sv_Johansson_Blood_Xmales)

# Change sentrixt position into factor
Covariates_sv_Johansson_Blood_Xfemales$sentrix_position<-factor(Covariates_sv_Johansson_Blood_Xfemales$sentrix_position)
Covariates_sv_Johansson_Blood_Xmales$sentrix_position<-factor(Covariates_sv_Johansson_Blood_Xmales$sentrix_position)
```

```{r}
#### Construct a Function called runDGLM for Xfemales and Xmales separately.
## Identify aDMCs in Johansson Blood females.
runDGLM_Johansson_Blood_Xfemales <- function(x)
{
  dat_Xfemales_dglm <- matrix(nrow=nrow(x), ncol=9)
  colnames(dat_Xfemales_dglm)<-c("cgID","effectsize.linear","std.error.linear","t.linear","p.linear","effectsize.disp","std.error.disp","t.disp","p.disp")
  for(CpG in 1:nrow(x))
  {
    print(CpG) #to check progress
    probe <- rownames(x)[CpG]
    tempdata <- data.frame(Age=Covariates_sv_Johansson_Blood_Xfemales[,2],data = x[probe,],sentrix_position=Covariates_sv_Johansson_Blood_Xfemales[,4],CD8T=Covariates_sv_Johansson_Blood_Xfemales[,5], CD4T=Covariates_sv_Johansson_Blood_Xfemales[,6], NK=Covariates_sv_Johansson_Blood_Xfemales[,7], Bcell=Covariates_sv_Johansson_Blood_Xfemales[,8], Mono=Covariates_sv_Johansson_Blood_Xfemales[,9],  SV1=Covariates_sv_Johansson_Blood_Xfemales[,11], SV2=Covariates_sv_Johansson_Blood_Xfemales[,12], SV3=Covariates_sv_Johansson_Blood_Xfemales[,13], SV4=Covariates_sv_Johansson_Blood_Xfemales[,14], SV5=Covariates_sv_Johansson_Blood_Xfemales[,15])
    
    fit_Xfemales <- tryCatch({ dglm(formula = data ~ Age + CD8T + CD4T + NK + Bcell + Mono + sentrix_position + SV1 + SV2 + SV3 + SV4 + SV5, dformula = ~ Age, data=tempdata, family = gaussian(link = identity))},error=identity)
    if(is.null(fit_Xfemales$message)){ #good
      #extract info about the linear model
      effectsize.linear <- summary(fit_Xfemales)$coefficients[2,1]
      std.error.linear<-summary(fit_Xfemales)$coefficients[2,2]
      t.value.linear <-summary(fit_Xfemales)$coefficients[2,3]
      p.value.linear <- summary(fit_Xfemales)$coefficients[2,4]
      
      #extract info about the dispersion model
      effectsize.disp <- summary(fit_Xfemales$dispersion.fit)$coefficients[2]
      std.error.disp<-summary(fit_Xfemales$dispersion.fit)$coefficients[4]
      t.value.disp <- summary(fit_Xfemales$dispersion.fit)$coefficients[6]
      p.value.disp <- summary(fit_Xfemales$dispersion.fit)$coefficients[8]
      
      # Summary statistics Output
      out <- matrix(c(probe,effectsize.linear,std.error.linear,t.value.linear,p.value.linear,effectsize.disp,std.error.disp,t.value.disp,p.value.disp), ncol=9)
      dat_Xfemales_dglm[CpG,] <- out
    }else{                  #bad
      print("Didn't converge")
      out <- matrix(c(probe, fit_Xfemales$message, rep(NA, times = 7)), ncol=9)
      dat_Xfemales_dglm[CpG,] <- out
    }
    
  }
  return(dat_Xfemales_dglm)
}

# Run runDGLM function
DGLM_Johansson_Blood_Xfemales<- runDGLM_Johansson_Blood_Xfemales(RIN.mvalues_Johansson_Blood_Xfemales)
DGLM_Johansson_Blood_Xfemales<-data.frame(DGLM_Johansson_Blood_Xfemales) 
rownames(DGLM_Johansson_Blood_Xfemales)<-make.names(DGLM_Johansson_Blood_Xfemales$cgID,unique=TRUE)
DGLM_Johansson_Blood_Xfemales<-DGLM_Johansson_Blood_Xfemales[,2:9]
DGLM_Johansson_Blood_Xfemales_errors<-subset(DGLM_Johansson_Blood_Xfemales,is.na(DGLM_Johansson_Blood_Xfemales$t.linear)) # No errors(all CpGs converge)
DGLM_Johansson_Blood_Xfemales_converge<-DGLM_Johansson_Blood_Xfemales

# Convert each column to numerical variable.
DGLM_Johansson_Blood_Xfemales_converge<-as.data.frame(lapply(DGLM_Johansson_Blood_Xfemales_converge,as.numeric)) 
rownames(DGLM_Johansson_Blood_Xfemales_converge)<-make.names(rownames(DGLM_Johansson_Blood_Xfemales),unique=TRUE) 
save(DGLM_Johansson_Blood_Xfemales_converge,file = "Output_05_DGLM_Johansson_Blood_Xfemales_converge.RData")
```

```{r}
## Correct bias and inflation for linear model.
# Xfemales
set.seed(1)
bc_DGLM_Johansson_Blood_Xfemales<- bacon(DGLM_Johansson_Blood_Xfemales_converge[,3])

# Calculate bias and inflation
bias(bc_DGLM_Johansson_Blood_Xfemales)         # -0.11  
inflation(bc_DGLM_Johansson_Blood_Xfemales)    # 1.8

# Save histogram plot of test statistics 
tiff("Output_05_Bacon_DGLM_Johansson_Blood_Xfemales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Johansson_Blood_Xfemales,xlab="t-statistics",main="aDMCs in Johansson Blood females",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Johansson_Blood_Xfemales<-tstat(bc_DGLM_Johansson_Blood_Xfemales)
pvals_DGLM_Johansson_Blood_Xfemales <-pval(bc_DGLM_Johansson_Blood_Xfemales)
table(pvals_DGLM_Johansson_Blood_Xfemales<0.05)   # 1120  
rownames(pvals_DGLM_Johansson_Blood_Xfemales)<-make.names(rownames(DGLM_Johansson_Blood_Xfemales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Johansson_Blood_Xfemales_es<-bacon(NULL,DGLM_Johansson_Blood_Xfemales_converge[,1],DGLM_Johansson_Blood_Xfemales_converge[,2])
es_DGLM_Johansson_Blood_Xfemales<-es(bc_DGLM_Johansson_Blood_Xfemales_es)

# Summary Bacon statistics
dat_DGLM_Johansson_Blood_Xfemales_bacon_linear<-as.data.frame(cbind(es_DGLM_Johansson_Blood_Xfemales,tstats_DGLM_Johansson_Blood_Xfemales,pvals_DGLM_Johansson_Blood_Xfemales))
colnames(dat_DGLM_Johansson_Blood_Xfemales_bacon_linear)<-c("bacon_effectsize_linear","bacon_tstatistics_linear","bacon_pval_linear")
rownames(dat_DGLM_Johansson_Blood_Xfemales_bacon_linear)<-make.names(rownames(DGLM_Johansson_Blood_Xfemales_converge),unique = TRUE)

# check females effecsize of aDMCs between BIOS and Hannum 
dat_Johansson_Blood_Xfemales_aDMCs<-dat_DGLM_Johansson_Blood_Xfemales_bacon_linear[rownames(df_aDMCs_effectsize_BIOS),]
df_Johansson_validation_females<-cbind(dat_Johansson_Blood_Xfemales_aDMCs[,1],df_aDMCs_effectsize_BIOS[,-2])
colnames(df_Johansson_validation_females)<-c("Johansson","BIOS","symbol")

tiff("Output_05_Johansson_Validation of female effect size.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Johansson_validation_females) +
    aes(x = BIOS,y = Johansson,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Johansson Blood")
dev.off()
```

```{r}
#### Construct a Function called runDGLM for Xfemales and Xmales separately.
## Identify aDMCs in Johansson Blood males.
runDGLM_Johansson_Blood_Xmales <- function(x)
{
  dat_Xmales_dglm <- matrix(nrow=nrow(x), ncol=9)
  colnames(dat_Xmales_dglm)<-c("cgID","effectsize.linear","std.error.linear","t.linear","p.linear","effectsize.disp","std.error.disp","t.disp","p.disp")
  for(CpG in 1:nrow(x))
  {
    print(CpG) #to check progress
    probe <- rownames(x)[CpG]
    tempdata <- data.frame(Age=Covariates_sv_Johansson_Blood_Xmales[,2],data = x[probe,],sentrix_position=Covariates_sv_Johansson_Blood_Xmales[,4],CD8T=Covariates_sv_Johansson_Blood_Xmales[,5], CD4T=Covariates_sv_Johansson_Blood_Xmales[,6], NK=Covariates_sv_Johansson_Blood_Xmales[,7], Bcell=Covariates_sv_Johansson_Blood_Xmales[,8], Mono=Covariates_sv_Johansson_Blood_Xmales[,9],  SV1=Covariates_sv_Johansson_Blood_Xmales[,11], SV2=Covariates_sv_Johansson_Blood_Xmales[,12], SV3=Covariates_sv_Johansson_Blood_Xmales[,13], SV4=Covariates_sv_Johansson_Blood_Xmales[,14], SV5=Covariates_sv_Johansson_Blood_Xmales[,15])
    
    fit_Xmales <- tryCatch({ dglm(formula = data ~ Age + CD8T + CD4T + NK + Bcell + Mono + sentrix_position + SV1 + SV2 + SV3 + SV4 + SV5, dformula = ~ Age, data=tempdata, family = gaussian(link = identity))},error=identity)
    if(is.null(fit_Xmales$message)){ #good
      #extract info about the linear model
      effectsize.linear <- summary(fit_Xmales)$coefficients[2,1]
      std.error.linear<-summary(fit_Xmales)$coefficients[2,2]
      t.value.linear <-summary(fit_Xmales)$coefficients[2,3]
      p.value.linear <- summary(fit_Xmales)$coefficients[2,4]
      
      #extract info about the dispersion model
      effectsize.disp <- summary(fit_Xmales$dispersion.fit)$coefficients[2]
      std.error.disp<-summary(fit_Xmales$dispersion.fit)$coefficients[4]
      t.value.disp <- summary(fit_Xmales$dispersion.fit)$coefficients[6]
      p.value.disp <- summary(fit_Xmales$dispersion.fit)$coefficients[8]
      
      # Summary statistics Output
      out <- matrix(c(probe,effectsize.linear,std.error.linear,t.value.linear,p.value.linear,effectsize.disp,std.error.disp,t.value.disp,p.value.disp), ncol=9)
      dat_Xmales_dglm[CpG,] <- out
    }else{                  #bad
      print("Didn't converge")
      out <- matrix(c(probe, fit_Xmales$message, rep(NA, times = 7)), ncol=9)
      dat_Xmales_dglm[CpG,] <- out
    }
    
  }
  return(dat_Xmales_dglm)
}

# Run runDGLM function
DGLM_Johansson_Blood_Xmales<- runDGLM_Johansson_Blood_Xmales(RIN.mvalues_Johansson_Blood_Xmales)
DGLM_Johansson_Blood_Xmales<-data.frame(DGLM_Johansson_Blood_Xmales) 
rownames(DGLM_Johansson_Blood_Xmales)<-make.names(DGLM_Johansson_Blood_Xmales$cgID,unique=TRUE)
DGLM_Johansson_Blood_Xmales<-DGLM_Johansson_Blood_Xmales[,2:9]
DGLM_Johansson_Blood_Xmales_errors<-subset(DGLM_Johansson_Blood_Xmales,is.na(DGLM_Johansson_Blood_Xmales$t.linear)) # No errors(all CpGs converge)
DGLM_Johansson_Blood_Xmales_converge<-DGLM_Johansson_Blood_Xmales

# Convert each column to numerical variable.
DGLM_Johansson_Blood_Xmales_converge<-as.data.frame(lapply(DGLM_Johansson_Blood_Xmales_converge,as.numeric)) 
rownames(DGLM_Johansson_Blood_Xmales_converge)<-make.names(rownames(DGLM_Johansson_Blood_Xmales),unique=TRUE)
save(DGLM_Johansson_Blood_Xmales_converge,file = "Output_05_DGLM_Johansson_Blood_Xmales_converge.RData")
```

```{r}
## Correct bias and inflation for linear model.
# Xmales
set.seed(1)
bc_DGLM_Johansson_Blood_Xmales<- bacon(DGLM_Johansson_Blood_Xmales_converge[,3])

# Calculate bias and inflation
bias(bc_DGLM_Johansson_Blood_Xmales)         # -0.07  
inflation(bc_DGLM_Johansson_Blood_Xmales)    # 1.2

# Save histogram plot of test statistics 
tiff("Output_05_Bacon_DGLM_Johansson_Blood_Xmales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Johansson_Blood_Xmales,xlab="t-statistics",main="aDMCs in Johansson Blood males",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Johansson_Blood_Xmales<-tstat(bc_DGLM_Johansson_Blood_Xmales)
pvals_DGLM_Johansson_Blood_Xmales <-pval(bc_DGLM_Johansson_Blood_Xmales)
table(pvals_DGLM_Johansson_Blood_Xmales<0.05)   # 3169  
rownames(pvals_DGLM_Johansson_Blood_Xmales)<-make.names(rownames(DGLM_Johansson_Blood_Xmales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Johansson_Blood_Xmales_es<-bacon(NULL,DGLM_Johansson_Blood_Xmales_converge[,1],DGLM_Johansson_Blood_Xmales_converge[,2])
es_DGLM_Johansson_Blood_Xmales<-es(bc_DGLM_Johansson_Blood_Xmales_es)

# Summary Bacon statistics
dat_DGLM_Johansson_Blood_Xmales_bacon_linear<-as.data.frame(cbind(es_DGLM_Johansson_Blood_Xmales,tstats_DGLM_Johansson_Blood_Xmales,pvals_DGLM_Johansson_Blood_Xmales))
colnames(dat_DGLM_Johansson_Blood_Xmales_bacon_linear)<-c("bacon_effectsize_linear","bacon_tstatistics_linear","bacon_pval_linear")
rownames(dat_DGLM_Johansson_Blood_Xmales_bacon_linear)<-make.names(rownames(DGLM_Johansson_Blood_Xmales_converge),unique = TRUE)

# check males effecsize of aDMCs between BIOS and Hannum 
dat_Johansson_Blood_Xmales_aDMCs<-dat_DGLM_Johansson_Blood_Xmales_bacon_linear[rownames(df_aDMCs_effectsize_BIOS),]
df_Johansson_validation_males<-cbind(dat_Johansson_Blood_Xmales_aDMCs[,1],df_aDMCs_effectsize_BIOS[,-1])
colnames(df_Johansson_validation_males)<-c("Johansson","BIOS","symbol")

tiff("Output_05_Johansson_Validation of male effect size.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Johansson_validation_males) +
    aes(x = BIOS,y = Johansson,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Johansson Blood")
dev.off()
```

```{r}
# save the validation data of effect size 
save(df_Johansson_validation_females,df_Johansson_validation_males,file="Output_05_Johansson_validation.RData")
```

```{r}
# Extract 66(34+32) female p-values and 1522(1490+32) male p-values.
dat_Johansson_Blood_Xfemales_aDMCs$symbol<-df_aDMCs_effectsize_BIOS$symbol
df_aDMCs_females_Johansson_Blood<-filter(dat_Johansson_Blood_Xfemales_aDMCs,symbol=="female-specific"|symbol=="sex-independent")
pvals_aDMCs_Xfemales_Johansson_Blood<-as.data.frame(pvals_DGLM_Johansson_Blood_Xfemales[rownames(df_aDMCs_females_Johansson_Blood),])
colnames(pvals_aDMCs_Xfemales_Johansson_Blood)<-"pval_females"

dat_Johansson_Blood_Xmales_aDMCs$symbol<-df_aDMCs_effectsize_BIOS$symbol
df_aDMCs_males_Johansson_Blood<-filter(dat_Johansson_Blood_Xmales_aDMCs,symbol=="male-specific"|symbol=="sex-independent")
pvals_aDMCs_Xmales_Johansson_Blood<-as.data.frame(pvals_DGLM_Johansson_Blood_Xmales[rownames(df_aDMCs_males_Johansson_Blood),])
colnames(pvals_aDMCs_Xmales_Johansson_Blood)<-"pval_males"

# Perform FDR correction for 1588 pvals(66+1522)
padjs_aDMCs_Johansson_Blood<-as.data.frame(p.adjust(c(pvals_aDMCs_Xfemales_Johansson_Blood$pval_females,pvals_aDMCs_Xmales_Johansson_Blood$pval_males),method="BH"))
colnames(padjs_aDMCs_Johansson_Blood)<-"bacon_p.adj_linear"

# check males (how many of 1522) and females (how many of 66) separately how many were FDR significant 
# 66 female aDMCs are still FDR significant in Hannum blood data.
padjs_aDMCs_females_Johansson_Blood<-as.data.frame(padjs_aDMCs_Johansson_Blood[1:66,])
colnames(padjs_aDMCs_females_Johansson_Blood)<-"bacon_p.adj_linear"
df_aDMCs_females_Johansson_Blood$bacon_p.adj_linear<-padjs_aDMCs_females_Johansson_Blood$bacon_p.adj_linear

df_aDMCs_females_Johansson_Blood<-filter(df_aDMCs_females_Johansson_Blood,bacon_p.adj_linear<0.05)
df_aDMCs_females_effectsize_up_Johansson<-filter(df_aDMCs_females_Johansson_Blood,bacon_effectsize_linear>0) # 35
df_aDMCs_females_effectsize_down_Johansson<-filter(df_aDMCs_females_Johansson_Blood,bacon_effectsize_linear<0) # 31

# 1508 male aDMCs are still FDR significant in Hannum blood data.
padjs_aDMCs_males_Johansson_Blood<-as.data.frame(padjs_aDMCs_Johansson_Blood[67:1588,])
colnames(padjs_aDMCs_males_Johansson_Blood)<-"bacon_p.adj_linear"
df_aDMCs_males_Johansson_Blood$bacon_p.adj_linear<-padjs_aDMCs_males_Johansson_Blood$bacon_p.adj_linear

df_aDMCs_males_Johansson_Blood<-filter(df_aDMCs_males_Johansson_Blood,bacon_p.adj_linear<0.05)
df_aDMCs_males_effectsize_up_Johansson<-filter(df_aDMCs_males_Johansson_Blood,bacon_effectsize_linear>0) # 965
df_aDMCs_males_effectsize_down_Johansson<-filter(df_aDMCs_males_Johansson_Blood,bacon_effectsize_linear<0) # 543

save(df_aDMCs_females_effectsize_up_Johansson,df_aDMCs_females_effectsize_down_Johansson,df_aDMCs_females_Johansson_Blood,df_aDMCs_males_effectsize_up_Johansson,df_aDMCs_males_effectsize_down_Johansson,df_aDMCs_males_Johansson_Blood,file = "Output_05_Johansson_Catogue of aDMCs.RData")
```

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```{r}
## aDMCs validation on Liu Mono datasets.
# Predict sex
predictedSex <- as.data.frame(getSex.DNAmArray(betas_Liu_Mono_chrX))
colnames(predictedSex)<-"predictedSex"
Covariates_Liu_Mono$predictedSex<-predictedSex$predictedSex
Covariates_Liu_Mono_Xfemales<-subset(Covariates_Liu_Mono,predictedSex=="Female")
Covariates_Liu_Mono_Xmales<-subset(Covariates_Liu_Mono,predictedSex=="Male")

# Split mvalues/betas object based on sex.
predictedSex_females<-subset(predictedSex,predictedSex=="Female")
predictedSex_males<-subset(predictedSex,predictedSex=="Male")
betas_Liu_Mono_Xfemales<-betas_Liu_Mono_chrX[,rownames(predictedSex_females)]
betas_Liu_Mono_Xmales<-betas_Liu_Mono_chrX[,rownames(predictedSex_males)]
mvalues_Liu_Mono_Xfemales<-mvalues_Liu_Mono_chrX[,rownames(predictedSex_females)]
mvalues_Liu_Mono_Xmales<-mvalues_Liu_Mono_chrX[,rownames(predictedSex_males)]

dim(betas_Liu_Mono_Xfemales)  # 9861  619
dim(betas_Liu_Mono_Xmales)    # 9861  583
dim(mvalues_Liu_Mono_Xfemales)  # 9861  619
dim(mvalues_Liu_Mono_Xmales)    # 9861  583

# we found 16 problematic "female" samples with strange DNAm,remove them(111,176,259,275,295,322,334,351,401,439,475,486,495,503,612,617) 
betas_Liu_Mono_Xfemales<-betas_Liu_Mono_Xfemales[,-c(111,176,259,275,295,322,334,351,401,439,475,486,495,503,612,617)]
mvalues_Liu_Mono_Xfemales<-mvalues_Liu_Mono_Xfemales[,-c(111,176,259,275,295,322,334,351,401,439,475,486,495,503,612,617)]
Covariates_Liu_Mono_Xfemales<-Covariates_Liu_Mono_Xfemales[-c(111,176,259,275,295,322,334,351,401,439,475,486,495,503,612,617),] 

dim(betas_Liu_Mono_Xfemales)  # 9861  603
dim(betas_Liu_Mono_Xmales)    # 9861  583
dim(mvalues_Liu_Mono_Xfemales)  # 9861  603
dim(mvalues_Liu_Mono_Xmales)    # 9861  583

# check female and male X methylation profile in Mono datasets.
tiff("Females X methylation in Mono.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
densityPlot(as.matrix(betas_Liu_Mono_Xfemales),main="Females X methylation in Mono", xlab="Beta values",xlim=c(0,1), cex.main=1.5, cex.lab=1.3, cex.axis=1.25) 
dev.off()

tiff("Males X methylation in Mono.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
densityPlot(as.matrix(betas_Liu_Mono_Xmales),main="Males X methylation in Mono", xlab="Beta values",xlim=c(0,1), cex.main=1.5, cex.lab=1.3, cex.axis=1.25) 
dev.off()

save(betas_Liu_Mono_Xfemales,betas_Liu_Mono_Xmales,mvalues_Liu_Mono_Xfemales,mvalues_Liu_Mono_Xmales,Covariates_Liu_Mono_Xfemales,Covariates_Liu_Mono_Xmales,file = "Output_05_Liu_Monocytes_chrX.RData")
```

```{r}
# rank-inverse normal(RIN) transformation for Xfemales and Xmales separately.
RIN.mvalues_Liu_Mono_Xfemales <- t(apply(mvalues_Liu_Mono_Xfemales, 1, RIN))
RIN.mvalues_Liu_Mono_Xmales <- t(apply(mvalues_Liu_Mono_Xmales, 1, RIN))
```

```{r}
# Create PCs of the Liu Mono data for Xfemales, Xmales.
pc_Liu_Mono_Xfemales<- prcomp_irlba(t(RIN.mvalues_Liu_Mono_Xfemales), n=10)
pc_Liu_Mono_Xmales<- prcomp_irlba(t(RIN.mvalues_Liu_Mono_Xmales), n=10)

# Screenplot drawing
summary(pc_Liu_Mono_Xfemales)
png("Output_05_screenplot_Liu_Mono_Xfemales.png")
screeplot(pc_Liu_Mono_Xfemales,type = "lines",main="Liu_Mono_Xfemales")
dev.off()

summary(pc_Liu_Mono_Xmales)
png("Output_05_screenplot_Liu_Mono_Xmales.png")
screeplot(pc_Liu_Mono_Xmales,type = "lines",main="Liu_Mono_Xmales")
dev.off()
```

```{r}
## Prepare SVA
# Liu Mono Xfemales
## null model:only includes the intercept
design0_Liu_Mono_Xfemales = model.matrix(~ 1,data=Covariates_Liu_Mono_Xfemales)
## Full model:all covarites
design_Liu_Mono_Xfemales <- model.matrix(~Age + Bcell + Tcell + NK + Neutro + sentrix_position, data=Covariates_Liu_Mono_Xfemales)

## Estimate Hannum female latent factors by SVA
svobj_Liu_Mono_Xfemales = sva(RIN.mvalues_Liu_Mono_Xfemales,design_Liu_Mono_Xfemales,design0_Liu_Mono_Xfemales,n.sv=5)
sv_Liu_Mono_Xfemales <- as.data.frame(svobj_Liu_Mono_Xfemales$sv)
colnames(sv_Liu_Mono_Xfemales) <- c('SV1', 'SV2', 'SV3', 'SV4', 'SV5')

# Liu Mono Xmales
## null model:only includes the intercept
design0_Liu_Mono_Xmales = model.matrix(~ 1,data=Covariates_Liu_Mono_Xmales)
## Full model:all covarites
design_Liu_Mono_Xmales <- model.matrix(~Age + Bcell + Tcell + NK + Neutro + sentrix_position, data=Covariates_Liu_Mono_Xmales)

# Estimate Hannum male latent factors by SVA
svobj_Liu_Mono_Xmales = sva(RIN.mvalues_Liu_Mono_Xmales,design_Liu_Mono_Xmales,design0_Liu_Mono_Xmales,n.sv=5)
sv_Liu_Mono_Xmales <- as.data.frame(svobj_Liu_Mono_Xmales$sv)
colnames(sv_Liu_Mono_Xmales) <- c('SV1', 'SV2', 'SV3', 'SV4', 'SV5')
```

```{r}
# Add estimated latent factors to covariates dataframe. 
#Xfemales
Covariates_sv_Liu_Mono_Xfemales<-cbind(Covariates_Liu_Mono_Xfemales,sv_Liu_Mono_Xfemales)

#Xmales
Covariates_sv_Liu_Mono_Xmales<-cbind(Covariates_Liu_Mono_Xmales,sv_Liu_Mono_Xmales)

# Change sentrixt position into factor
Covariates_sv_Liu_Mono_Xfemales$sentrix_position<-factor(Covariates_sv_Liu_Mono_Xfemales$sentrix_position)
Covariates_sv_Liu_Mono_Xmales$sentrix_position<-factor(Covariates_sv_Liu_Mono_Xmales$sentrix_position)
```

```{r}
#### Construct a Function called runDGLM for Xfemales and Xmales separately.
## Identify aDMCs in Liu Mono females.
runDGLM_Liu_Mono_Xfemales <- function(x)
{
  dat_Xfemales_dglm <- matrix(nrow=nrow(x), ncol=9)
  colnames(dat_Xfemales_dglm)<-c("cgID","effectsize.linear","std.error.linear","t.linear","p.linear","effectsize.disp","std.error.disp","t.disp","p.disp")
  for(CpG in 1:nrow(x))
  {
    print(CpG) #to check progress
    probe <- rownames(x)[CpG]
    tempdata <- data.frame(Age=Covariates_sv_Liu_Mono_Xfemales[,2],data = x[probe,],sentrix_position=Covariates_sv_Liu_Mono_Xfemales[,3],Bcell=Covariates_sv_Liu_Mono_Xfemales[,4],Tcell=Covariates_sv_Liu_Mono_Xfemales[,5], NK=Covariates_sv_Liu_Mono_Xfemales[,6], Neutro=Covariates_sv_Liu_Mono_Xfemales[,7],   SV1=Covariates_sv_Liu_Mono_Xfemales[,9], SV2=Covariates_sv_Liu_Mono_Xfemales[,10], SV3=Covariates_sv_Liu_Mono_Xfemales[,11], SV4=Covariates_sv_Liu_Mono_Xfemales[,12], SV5=Covariates_sv_Liu_Mono_Xfemales[,13])
    
    fit_Xfemales <- tryCatch({ dglm(formula = data ~ Age + Bcell + Tcell + NK + Neutro + sentrix_position + SV1 + SV2 + SV3 + SV4 + SV5, dformula = ~ Age, data=tempdata, family = gaussian(link = identity))},error=identity)
    if(is.null(fit_Xfemales$message)){ #good
      #extract info about the linear model
      effectsize.linear <- summary(fit_Xfemales)$coefficients[2,1]
      std.error.linear<-summary(fit_Xfemales)$coefficients[2,2]
      t.value.linear <-summary(fit_Xfemales)$coefficients[2,3]
      p.value.linear <- summary(fit_Xfemales)$coefficients[2,4]
      
      #extract info about the dispersion model
      effectsize.disp <- summary(fit_Xfemales$dispersion.fit)$coefficients[2]
      std.error.disp<-summary(fit_Xfemales$dispersion.fit)$coefficients[4]
      t.value.disp <- summary(fit_Xfemales$dispersion.fit)$coefficients[6]
      p.value.disp <- summary(fit_Xfemales$dispersion.fit)$coefficients[8]
      
      # Summary statistics Output
      out <- matrix(c(probe,effectsize.linear,std.error.linear,t.value.linear,p.value.linear,effectsize.disp,std.error.disp,t.value.disp,p.value.disp), ncol=9)
      dat_Xfemales_dglm[CpG,] <- out
    }else{                  #bad
      print("Didn't converge")
      out <- matrix(c(probe, fit_Xfemales$message, rep(NA, times = 7)), ncol=9)
      dat_Xfemales_dglm[CpG,] <- out
    }
    
  }
  return(dat_Xfemales_dglm)
}

# Run runDGLM function
DGLM_Liu_Mono_Xfemales<- runDGLM_Liu_Mono_Xfemales(RIN.mvalues_Liu_Mono_Xfemales)
DGLM_Liu_Mono_Xfemales<-data.frame(DGLM_Liu_Mono_Xfemales) 
rownames(DGLM_Liu_Mono_Xfemales)<-make.names(DGLM_Liu_Mono_Xfemales$cgID,unique=TRUE)
DGLM_Liu_Mono_Xfemales<-DGLM_Liu_Mono_Xfemales[,2:9]
DGLM_Liu_Mono_Xfemales_errors<-subset(DGLM_Liu_Mono_Xfemales,is.na(DGLM_Liu_Mono_Xfemales$t.linear)) # No errors(all CpGs converge)
DGLM_Liu_Mono_Xfemales_converge<-DGLM_Liu_Mono_Xfemales

# Convert each column to numerical variable.
DGLM_Liu_Mono_Xfemales_converge<-as.data.frame(lapply(DGLM_Liu_Mono_Xfemales_converge,as.numeric)) 
rownames(DGLM_Liu_Mono_Xfemales_converge)<-make.names(rownames(DGLM_Liu_Mono_Xfemales),unique=TRUE) 
save(DGLM_Liu_Mono_Xfemales_converge,file = "Output_05_DGLM_Liu_Mono_Xfemales_converge.RData")
```

```{r}
## Correct bias and inflation for linear model.
# Xfemales
set.seed(1)
bc_DGLM_Liu_Mono_Xfemales<- bacon(DGLM_Liu_Mono_Xfemales_converge[,3])

# Calculate bias and inflation
bias(bc_DGLM_Liu_Mono_Xfemales)         # -0.07  
inflation(bc_DGLM_Liu_Mono_Xfemales)    # 1.19

# Save histogram plot of test statistics 
tiff("Output_05_Bacon_DGLM_Liu_Mono_Xfemales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Liu_Mono_Xfemales,xlab="t-statistics",main="aDMCs in Liu Mono females",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Liu_Mono_Xfemales<-tstat(bc_DGLM_Liu_Mono_Xfemales)
pvals_DGLM_Liu_Mono_Xfemales <-pval(bc_DGLM_Liu_Mono_Xfemales)
table(pvals_DGLM_Liu_Mono_Xfemales<0.05)   # 714  
rownames(pvals_DGLM_Liu_Mono_Xfemales)<-make.names(rownames(DGLM_Liu_Mono_Xfemales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Liu_Mono_Xfemales_es<-bacon(NULL,DGLM_Liu_Mono_Xfemales_converge[,1],DGLM_Liu_Mono_Xfemales_converge[,2])
es_DGLM_Liu_Mono_Xfemales<-es(bc_DGLM_Liu_Mono_Xfemales_es)

# Summary Bacon statistics
dat_DGLM_Liu_Mono_Xfemales_bacon_linear<-as.data.frame(cbind(es_DGLM_Liu_Mono_Xfemales,tstats_DGLM_Liu_Mono_Xfemales,pvals_DGLM_Liu_Mono_Xfemales))
colnames(dat_DGLM_Liu_Mono_Xfemales_bacon_linear)<-c("bacon_effectsize_linear","bacon_tstatistics_linear","bacon_pval_linear")
rownames(dat_DGLM_Liu_Mono_Xfemales_bacon_linear)<-make.names(rownames(DGLM_Liu_Mono_Xfemales_converge),unique = TRUE)

# check females effecsize of aDMCs between BIOS and Liu Mono 
dat_Liu_Mono_Xfemales_aDMCs<-dat_DGLM_Liu_Mono_Xfemales_bacon_linear[rownames(df_aDMCs_effectsize_BIOS),]
df_Liu_validation_females<-cbind(dat_Liu_Mono_Xfemales_aDMCs[,1],df_aDMCs_effectsize_BIOS[,-2])
colnames(df_Liu_validation_females)<-c("Liu","BIOS","symbol")

tiff("Output_05_Liu_Validation of female size effect.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Liu_validation_females) +
    aes(x = BIOS,y = Liu,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Liu Monocytes")
dev.off()
```

```{r}
#### Construct a Function called runDGLM for Xfemales and Xmales separately.
## Identify aDMCs in Liu Mono males.
runDGLM_Liu_Mono_Xmales <- function(x)
{
  dat_Xmales_dglm <- matrix(nrow=nrow(x), ncol=9)
  colnames(dat_Xmales_dglm)<-c("cgID","effectsize.linear","std.error.linear","t.linear","p.linear","effectsize.disp","std.error.disp","t.disp","p.disp")
  for(CpG in 1:nrow(x))
  {
    print(CpG) #to check progress
    probe <- rownames(x)[CpG]
    tempdata <- data.frame(Age=Covariates_sv_Liu_Mono_Xmales[,2],data = x[probe,],sentrix_position=Covariates_sv_Liu_Mono_Xmales[,3],Bcell=Covariates_sv_Liu_Mono_Xmales[,4],Tcell=Covariates_sv_Liu_Mono_Xmales[,5], NK=Covariates_sv_Liu_Mono_Xmales[,6],Neutro=Covariates_sv_Liu_Mono_Xmales[,7],SV1=Covariates_sv_Liu_Mono_Xmales[,9],SV2=Covariates_sv_Liu_Mono_Xmales[,10],SV3=Covariates_sv_Liu_Mono_Xmales[,11],SV4=Covariates_sv_Liu_Mono_Xmales[,12],SV5=Covariates_sv_Liu_Mono_Xmales[,13])
    
    fit_Xmales <- tryCatch({ dglm(formula = data ~ Age + Bcell + Tcell + NK + Neutro + sentrix_position + SV1 + SV2 + SV3 + SV4 + SV5, dformula = ~ Age, data=tempdata, family = gaussian(link = identity))},error=identity)
    if(is.null(fit_Xmales$message)){ #good
      #extract info about the linear model
      effectsize.linear <- summary(fit_Xmales)$coefficients[2,1]
      std.error.linear<-summary(fit_Xmales)$coefficients[2,2]
      t.value.linear <-summary(fit_Xmales)$coefficients[2,3]
      p.value.linear <- summary(fit_Xmales)$coefficients[2,4]
      
      #extract info about the dispersion model
      effectsize.disp <- summary(fit_Xmales$dispersion.fit)$coefficients[2]
      std.error.disp<-summary(fit_Xmales$dispersion.fit)$coefficients[4]
      t.value.disp <- summary(fit_Xmales$dispersion.fit)$coefficients[6]
      p.value.disp <- summary(fit_Xmales$dispersion.fit)$coefficients[8]
      
      # Summary statistics Output
      out <- matrix(c(probe,effectsize.linear,std.error.linear,t.value.linear,p.value.linear,effectsize.disp,std.error.disp,t.value.disp,p.value.disp), ncol=9)
      dat_Xmales_dglm[CpG,] <- out
    }else{                  #bad
      print("Didn't converge")
      out <- matrix(c(probe, fit_Xmales$message, rep(NA, times = 7)), ncol=9)
      dat_Xmales_dglm[CpG,] <- out
    }
    
  }
  return(dat_Xmales_dglm)
}

# Run runDGLM function
DGLM_Liu_Mono_Xmales<- runDGLM_Liu_Mono_Xmales(RIN.mvalues_Liu_Mono_Xmales)
DGLM_Liu_Mono_Xmales<-data.frame(DGLM_Liu_Mono_Xmales) 
rownames(DGLM_Liu_Mono_Xmales)<-make.names(DGLM_Liu_Mono_Xmales$cgID,unique=TRUE)
DGLM_Liu_Mono_Xmales<-DGLM_Liu_Mono_Xmales[,2:9]
DGLM_Liu_Mono_Xmales_errors<-subset(DGLM_Liu_Mono_Xmales,is.na(DGLM_Liu_Mono_Xmales$t.linear)) # No errors(all CpGs converge)
DGLM_Liu_Mono_Xmales_converge<-DGLM_Liu_Mono_Xmales

# Convert each column to numerical variable.
DGLM_Liu_Mono_Xmales_converge<-as.data.frame(lapply(DGLM_Liu_Mono_Xmales_converge,as.numeric)) 
rownames(DGLM_Liu_Mono_Xmales_converge)<-make.names(rownames(DGLM_Liu_Mono_Xmales),unique=TRUE) 
save(DGLM_Liu_Mono_Xmales_converge,file = "Output_05_DGLM_Liu_Mono_Xmales_converge.RData")
```

```{r}
## Correct bias and inflation for linear model.
# Xmales
set.seed(1)
bc_DGLM_Liu_Mono_Xmales<- bacon(DGLM_Liu_Mono_Xmales_converge[,3])

# Calculate bias and inflation
bias(bc_DGLM_Liu_Mono_Xmales)         # -0.01  
inflation(bc_DGLM_Liu_Mono_Xmales)    # 1.16

# Save histogram plot of test statistics 
tiff("Output_05_Bacon_DGLM_Liu_Mono_Xmales.tiff", units="in", width=8, height=6, res=600,compression = 'lzw')
fit(bc_DGLM_Liu_Mono_Xmales,xlab="t-statistics",main="aDMCs in Liu Mono males",n=100)
dev.off()

#Extract the BACON-adjusted t-statistics and p-values
tstats_DGLM_Liu_Mono_Xmales<-tstat(bc_DGLM_Liu_Mono_Xmales)
pvals_DGLM_Liu_Mono_Xmales <-pval(bc_DGLM_Liu_Mono_Xmales)
table(pvals_DGLM_Liu_Mono_Xmales<0.05)   # 972  
rownames(pvals_DGLM_Liu_Mono_Xmales)<-make.names(rownames(DGLM_Liu_Mono_Xmales_converge),unique=TRUE) 

# Extract BACON-adjusted effectsize.
set.seed(1)
bc_DGLM_Liu_Mono_Xmales_es<-bacon(NULL,DGLM_Liu_Mono_Xmales_converge[,1],DGLM_Liu_Mono_Xmales_converge[,2])
es_DGLM_Liu_Mono_Xmales<-es(bc_DGLM_Liu_Mono_Xmales_es)

# Summary Bacon statistics
dat_DGLM_Liu_Mono_Xmales_bacon_linear<-as.data.frame(cbind(es_DGLM_Liu_Mono_Xmales,tstats_DGLM_Liu_Mono_Xmales,pvals_DGLM_Liu_Mono_Xmales))
colnames(dat_DGLM_Liu_Mono_Xmales_bacon_linear)<-c("bacon_effectsize_linear","bacon_tstatistics_linear","bacon_pval_linear")
rownames(dat_DGLM_Liu_Mono_Xmales_bacon_linear)<-make.names(rownames(DGLM_Liu_Mono_Xmales_converge),unique = TRUE)

# check females effecsize of aDMCs between BIOS and Liu Mono 
dat_Liu_Mono_Xmales_aDMCs<-dat_DGLM_Liu_Mono_Xmales_bacon_linear[rownames(df_aDMCs_effectsize_BIOS),]
df_Liu_validation_males<-cbind(dat_Liu_Mono_Xmales_aDMCs[,1],df_aDMCs_effectsize_BIOS[,-1])
colnames(df_Liu_validation_males)<-c("Liu","BIOS","symbol")

tiff("Output_05_Liu_Validation of male size effect.tiff", units="in", width=12, height=8, res=800,compression = 'lzw')
ggplot(df_Liu_validation_males) +
    aes(x = BIOS,y = Liu,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Liu Monocytes")
dev.off()
```

```{r}
# save the validation data of effect size 
save(df_Liu_validation_females,df_Liu_validation_males,file="Output_05_Liu_validation.RData")
```

```{r}
# Extract 66(34+32) female p-values and 1522(1490+32) male p-values.
dat_Liu_Mono_Xfemales_aDMCs$symbol<-df_aDMCs_effectsize_BIOS$symbol
df_aDMCs_females_Liu_Mono<-filter(dat_Liu_Mono_Xfemales_aDMCs,symbol=="female-specific"|symbol=="sex-independent")
pvals_aDMCs_Xfemales_Liu_Mono<-as.data.frame(pvals_DGLM_Liu_Mono_Xfemales[rownames(df_aDMCs_females_Liu_Mono),])
colnames(pvals_aDMCs_Xfemales_Liu_Mono)<-"pval_females"

dat_Liu_Mono_Xmales_aDMCs$symbol<-df_aDMCs_effectsize_BIOS$symbol
df_aDMCs_males_Liu_Mono<-filter(dat_Liu_Mono_Xmales_aDMCs,symbol=="male-specific"|symbol=="sex-independent")
pvals_aDMCs_Xmales_Liu_Mono<-as.data.frame(pvals_DGLM_Liu_Mono_Xmales[rownames(df_aDMCs_males_Liu_Mono),])
colnames(pvals_aDMCs_Xmales_Liu_Mono)<-"pval_males"

# Perform FDR correction for 1588 pvals(66+1522)
padjs_aDMCs_Liu_Mono<-as.data.frame(p.adjust(c(pvals_aDMCs_Xfemales_Liu_Mono$pval_females,pvals_aDMCs_Xmales_Liu_Mono$pval_males),method="BH"))
colnames(padjs_aDMCs_Liu_Mono)<-"bacon_p.adj_linear"

# check males (how many of 1522) and females (how many of 66) separately how many were FDR significant 
# 30 female aDMCs are still FDR significant in Liu Mono data.
padjs_aDMCs_females_Liu_Mono<-as.data.frame(padjs_aDMCs_Liu_Mono[1:66,])
colnames(padjs_aDMCs_females_Liu_Mono)<-"bacon_p.adj_linear"
df_aDMCs_females_Liu_Mono$bacon_p.adj_linear<-padjs_aDMCs_females_Liu_Mono$bacon_p.adj_linear

df_aDMCs_females_Liu_Mono<-filter(df_aDMCs_females_Liu_Mono,bacon_p.adj_linear<0.05)
df_aDMCs_females_effectsize_up_Liu<-filter(df_aDMCs_females_Liu_Mono,bacon_effectsize_linear>0) # 10
df_aDMCs_females_effectsize_down_Liu<-filter(df_aDMCs_females_Liu_Mono,bacon_effectsize_linear<0) # 20

# 280 male aDMCs are still FDR significant in Hannum blood data.
padjs_aDMCs_males_Liu_Mono<-as.data.frame(padjs_aDMCs_Liu_Mono[67:1588,])
colnames(padjs_aDMCs_males_Liu_Mono)<-"bacon_p.adj_linear"
df_aDMCs_males_Liu_Mono$bacon_p.adj_linear<-padjs_aDMCs_males_Liu_Mono$bacon_p.adj_linear

df_aDMCs_males_Liu_Mono<-filter(df_aDMCs_males_Liu_Mono,bacon_p.adj_linear<0.05)
df_aDMCs_males_effectsize_up_Liu<-filter(df_aDMCs_males_Liu_Mono,bacon_effectsize_linear>0) # 149
df_aDMCs_males_effectsize_down_Liu<-filter(df_aDMCs_males_Liu_Mono,bacon_effectsize_linear<0) # 131

save(df_aDMCs_females_effectsize_up_Liu,df_aDMCs_females_effectsize_down_Liu,df_aDMCs_females_Liu_Mono,df_aDMCs_males_effectsize_up_Liu,df_aDMCs_males_effectsize_down_Liu,df_aDMCs_males_Liu_Mono,file = "Output_05_Liu_Catogue of aDMCs.RData")
```

```{r}
# The catlogue of validated aDMCs in females and males.
# Females:29
Overlap_females<-Reduce(intersect, list(rownames(dat_DGLM_Xfemales_bacon_linear),rownames(df_aDMCs_females_Johansson_Blood),rownames(df_aDMCs_females_Hannum_Blood),rownames(df_aDMCs_females_Liu_Mono)))
dat_aDMCs_females_validated<-dat_DGLM_Xfemales_bacon_linear[Overlap_females,]

# Males: 190
Overlap_males<-Reduce(intersect, list(rownames(dat_DGLM_Xmales_bacon_linear),rownames(df_aDMCs_males_Johansson_Blood),rownames(df_aDMCs_males_Hannum_Blood),rownames(df_aDMCs_males_Liu_Mono)))
dat_aDMCs_males_validated<-dat_DGLM_Xmales_bacon_linear[Overlap_males,]

# Statistically significant aDMCs both in females and males.
both_sex<-intersect(rownames(dat_aDMCs_females_validated),rownames(dat_aDMCs_males_validated))
df_aDMCs_both_sex_validated<-dat_DGLM_Xfemales_bacon_linear[both_sex,]

# Statistically significant aDMCs only in females.
df_aDMCs_female_specific_validated<-subset(dat_aDMCs_females_validated,!rownames(dat_aDMCs_females_validated)%in%rownames(df_aDMCs_both_sex_validated))

# Statistically significant aDMCs only in males.
df_aDMCs_male_specific_validated<-subset(dat_aDMCs_males_validated,!rownames(dat_aDMCs_males_validated)%in% both_sex)

save(df_aDMCs_both_sex_validated,df_aDMCs_female_specific_validated,df_aDMCs_male_specific_validated,dat_aDMCs_females_validated,dat_aDMCs_males_validated,file = "Output_05_Catalogue_aDMCs_validated.RData")
```

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```{r}
tiff("Output_05_Validation of aDMCs.tiff", units="in", width=18, height=14, res=800,compression = 'lzw')
p1<-ggplot(df_Johansson_validation_females) +
    aes(x = BIOS,y = Johansson,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Johansson Blood")

p2<-ggplot(df_Johansson_validation_males) +
    aes(x = BIOS,y = Johansson,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Johansson Blood")

p3<-ggplot(df_Hannum_validation_females) +
    aes(x = BIOS,y = Hannum,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Hannum Blood")

p4<-ggplot(df_Hannum_validation_males) +
    aes(x = BIOS,y = Hannum,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effectsize in BIOS",y="effectsize in Hannum Blood")

p5<-ggplot(df_Liu_validation_females) +
    aes(x = BIOS,y = Liu,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Females")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Liu Monocytes")

p6<-ggplot(df_Liu_validation_males) +
    aes(x = BIOS,y = Liu,color=symbol) +
     geom_point(size = 2, alpha = 0.5) + 
     geom_vline(aes(xintercept=0)) +
     geom_hline(aes(yintercept=0)) +
     geom_abline(slope=1, intercept = 0)+
  scale_x_continuous(limits=c(-0.04, 0.04))+
  scale_y_continuous(limits=c(-0.04, 0.04))+
  scale_color_discrete(name="aDMCs",breaks=c("female-specific","male-specific","sex-independent"),labels=c("female-specific:34CpGs","male-specific:1490CpGs","sex-independent:32CpGs"))+
  ggtitle("Males")+
  theme(plot.title = 
element_text(hjust = 0.5),text = element_text(size = 20,hjust = 0.5))+
    labs(x="effect size in BIOS",y="effect size in Liu Monocytes")
grid.arrange(p1,p2,p3,p4,p5,p6,ncol=2)
dev.off()
```

```{r}
# The overlap aDMCs between BIOS data and three external datasets.
# Venn plot
library(RColorBrewer)
myCol <- brewer.pal(4,"Pastel2")
display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}

# The statistically significant overlapping aDMCs in females
df_aDMCs_females<-list(Hannum=rownames(df_aDMCs_females_Hannum_Blood), Johansson =rownames(df_aDMCs_females_Johansson_Blood),Liu=rownames(df_aDMCs_females_Liu_Mono), BIOS=rownames(dat_DGLM_Xfemales_bacon_linear))

# The statistically significant overlapping aDMCs in males 
df_aDMCs_males<-list(Hannum=rownames(df_aDMCs_males_Hannum_Blood), Johansson =rownames(df_aDMCs_males_Johansson_Blood),Liu=rownames(df_aDMCs_males_Liu_Mono), BIOS=rownames(dat_DGLM_Xmales_bacon_linear))

# Venn plot.
tiff("Output_05_venn_aDMCs_females.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aDMCs_females,fill = myCol,cat.fontface = "bold",main="Females",main.fontface="bold",cex = 1.5,main.cex = 1.5,cat.cex=1.5)
dev.off()

tiff("Output_05_venn_aDMCs_males.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aDMCs_males,fill = myCol,cat.fontface = "bold",main="Males",main.fontface="bold",cex = 1.5,main.cex = 1.5,cat.cex=1.5)
dev.off()
```


------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```{r}
# The overlap aDMCs between BIOS data and three external datasets.
# Venn plot
library(RColorBrewer)
myCol <- brewer.pal(4,"Pastel2")
display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}

# The statistically significant overlapping aDMCs in females (effectsize<0).
df_aDMCs_females_effectsize_down<-list(Hannum=rownames(df_aDMCs_females_effectsize_down_Hannum), Johansson =rownames(df_aDMCs_females_effectsize_down_Johansson),Liu=rownames(df_aDMCs_females_effectsize_down_Liu), BIOS=rownames(dat_DGLM_Xfemales_bacon_linear_down))

tiff("Output_05_venn_aDMCs_effectsize_down_females.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aDMCs_females_effectsize_down,fill = myCol,cat.fontface = "bold",main="Statistically significant overlapping aDMCs in females(effect size<0)",main.fontface="bold")
dev.off()

# The statistically significant overlapping aDMCs in females (effectsize>0).
df_aDMCs_females_effectsize_up<-list(Hannum=rownames(df_aDMCs_females_effectsize_up_Hannum), Johansson =rownames(df_aDMCs_females_effectsize_up_Johansson),Liu=rownames(df_aDMCs_females_effectsize_up_Liu), BIOS=rownames(dat_DGLM_Xfemales_bacon_linear_up))

tiff("Output_05_venn_aDMCs_effectsize_up_females.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aDMCs_females_effectsize_up,fill = myCol,cat.fontface = "bold",main="Statistically significant overlapping aDMCs in females(effect size>0)",main.fontface="bold")
dev.off()

# The statistically significant overlapping aDMCs in males (effectsize<0).
df_aDMCs_males_effectsize_down<-list(Hannum=rownames(df_aDMCs_males_effectsize_down_Hannum), Johansson =rownames(df_aDMCs_males_effectsize_down_Johansson),Liu=rownames(df_aDMCs_males_effectsize_down_Liu), BIOS=rownames(dat_DGLM_Xmales_bacon_linear_down))

tiff("Output_05_venn_aDMCs_effectsize_down_males.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aDMCs_males_effectsize_down,fill = myCol,cat.fontface = "bold",main="Statistically significant overlapping aDMCs in males(effect size<0)",main.fontface="bold")
dev.off()

# The statistically significant overlapping aDMCs in males (effectsize>0).
df_aDMCs_males_effectsize_up<-list(Hannum=rownames(df_aDMCs_males_effectsize_up_Hannum), Johansson =rownames(df_aDMCs_males_effectsize_up_Johansson),Liu=rownames(df_aDMCs_males_effectsize_up_Liu), BIOS=rownames(dat_DGLM_Xmales_bacon_linear_up))

tiff("Output_05_venn_aDMCs_effectsize_up_males.tiff", units="in", width=9, height=8, res=800,compression ='lzw')
display_venn(df_aDMCs_males_effectsize_up,fill = myCol,cat.fontface = "bold",main="Statistically significant overlapping aDMCs in males(effect size>0)",main.fontface="bold")
dev.off()
```


































































































